{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# __R Session n¬∞5 : tutorial__\n",
    "M1 MEG UE5 - Claire Vandiedonck\n",
    "***\n",
    "\n",
    "## **tutorial on packages**\n",
    "\n",
    "*Content of this tutorial:*\n",
    "\n",
    "- lister des paquets install√©s\n",
    "- utiliser un paquet install√© utile en g√©nomique\n",
    "- installer un paquet utile en g√©nomique et l'utiliser\n",
    "\n",
    "L'exemple d'utilisation portera sur des variants g√©n√©tiques et des sites de chromatine ouverte."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## Avant d'aller plus loin\n",
    "\n",
    "<div class=\"alert alert-block alert-danger\"><b>Attention:</b> \n",
    "Ne travaillez pas directement sur ce notebook pour ne pas le perdre. Dupliquez-le et renommez-le par exemple en ajoutant vos initiales et travaillez sur cette nouvelle copie. Pour ce faire, dans le panneau de gauche, faites un clic droit sur le fichier et s√©lectionnez \"Duplicate\". Puis, toujours dans la colonne de gauche, faites un clic droit sur cette copie et s√©lectionnez \"rename\" pour changer le nom. Ouvrez ensuite cette nouvelle version en double cliquant dessus. Vous √™tes pr√™t(e) √† d√©marrer! <br>\n",
    "<br>\n",
    "<b>N'oubliez pas de sauvegarder r√©guli√®rement votre notebook</b>: <kbd>Ctrl</kbd> + <kbd>S</kbd>. ou en cliquant sur l'icone üíæ en haut √† gauche de votre notebook ou dans le Menu du JupyterLab \"File puis \"Save Notebook\"! Vous pouvez aussi le sauvegarder au format html: Menu \"File\" > Export Notebook As> Export notebook as HTML.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check your working directory and change it if needed:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'/srv/home/cvandiedonck/megm1_ue5_r/R5'"
      ],
      "text/latex": [
       "'/srv/home/cvandiedonck/megm1\\_ue5\\_r/R5'"
      ],
      "text/markdown": [
       "'/srv/home/cvandiedonck/megm1_ue5_r/R5'"
      ],
      "text/plain": [
       "[1] \"/srv/home/cvandiedonck/megm1_ue5_r/R5\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#cell 1\n",
    "getwd()\n",
    "#setwd('/srv/home/cvandiedonck/megm1_ue5_r/R5') #change with your login!!!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\"><b>Warning:</b> you are strongly advised to run the cells in the indicated order. If you want to rerun cells above, you can just restart the kernel to start at 1 again. </div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "---\n",
    "\n",
    "\n",
    "## 1. Lister les paquets install√©s\n",
    "---\n",
    "\n",
    "- identifier les r√©pertoires, c'est √† dire les `librairies` dans lesquels les paquets sont install√©s avec la function `.libPaths()`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'/srv/home/cvandiedonck/R/x86_64-conda-linux-gnu-library/4.0'</li><li>'/srv/conda/envs/notebook/lib/R/library'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item '/srv/home/cvandiedonck/R/x86\\_64-conda-linux-gnu-library/4.0'\n",
       "\\item '/srv/conda/envs/notebook/lib/R/library'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. '/srv/home/cvandiedonck/R/x86_64-conda-linux-gnu-library/4.0'\n",
       "2. '/srv/conda/envs/notebook/lib/R/library'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"/srv/home/cvandiedonck/R/x86_64-conda-linux-gnu-library/4.0\"\n",
       "[2] \"/srv/conda/envs/notebook/lib/R/library\"                     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#cell 2\n",
    ".libPaths()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\"> \n",
    "    \n",
    "<b><em> Info:</em></b> <br>\n",
    "Deux r√©pertoires apparaissent:<br>\n",
    "<b>- /srv/conda/envs/notebook/lib/R/library</b> qui correspond sur adenine au repertoire o√π les paquets ont √©t√© install√©s au moment de la cr√©ation de l'environnement<br>\n",
    "<b>- /srv/home/your_login/R/x86_64-conda-linux-gnu-library/4.0</b> sur votre home et qui contient tous les r√©pertoires que vous installez manuellement. Il contient notamment swirl que nous avions intall√© manuellement dans une pr√©c√©dente session.<br>\n",
    "\n",
    "De mani√®re g√©n√©rale, si vous travaillez pour la 1√®re fois sur une machine ou un serveur de calcul sans que vous ayez install√© manuellement un paquet, l'√©quivalent de ce second r√©pertoire sur votre home est absent.<br>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Combien de paquets sont install√©es dans ce(s) repertoire(s)?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "17"
      ],
      "text/latex": [
       "17"
      ],
      "text/markdown": [
       "17"
      ],
      "text/plain": [
       "[1] 17"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#cell 3\n",
    "length(list.files(.libPaths()[1]))# modifier l'indice selon le repertoire souhait√©"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Vous en aurez sans doute un peu moins que dans la version d√©j√† √©x√©cut√©e de ce notebook."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Identifier la liste des paquets install√©s dans ce(s) r√©pertoire(s).\n",
    "\n",
    "On pourrait utiliser la commande suivante:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'abind'</li><li>'AnnotationDbi'</li><li>'askpass'</li><li>'assertthat'</li><li>'backports'</li><li>'base'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'abind'\n",
       "\\item 'AnnotationDbi'\n",
       "\\item 'askpass'\n",
       "\\item 'assertthat'\n",
       "\\item 'backports'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'abind'\n",
       "2. 'AnnotationDbi'\n",
       "3. 'askpass'\n",
       "4. 'assertthat'\n",
       "5. 'backports'\n",
       "6. 'base'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"abind\"         \"AnnotationDbi\" \"askpass\"       \"assertthat\"   \n",
       "[5] \"backports\"     \"base\"         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#cell 4\n",
    "head(list.files(.libPaths()[2]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Mais on peut r√©cup√©rer plus d'information sur les versions et les repertoires avec la function `installed.packages()`. La fonction retournant un grand nombre d'informations, listons les avant d'en s√©lectionner les plus utiles."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " [1] \"Package\"               \"LibPath\"               \"Version\"              \n",
      " [4] \"Priority\"              \"Depends\"               \"Imports\"              \n",
      " [7] \"LinkingTo\"             \"Suggests\"              \"Enhances\"             \n",
      "[10] \"License\"               \"License_is_FOSS\"       \"License_restricts_use\"\n",
      "[13] \"OS_type\"               \"MD5sum\"                \"NeedsCompilation\"     \n",
      "[16] \"Built\"                \n"
     ]
    }
   ],
   "source": [
    "#cell 5\n",
    "print(colnames(installed.packages()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Affichons les 1√®re, 2√®me et 3√®me colonnes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 6 √ó 3 of type chr</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Package</th><th scope=col>LibPath</th><th scope=col>Version</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>beeswarm</th><td>beeswarm</td><td>/srv/home/cvandiedonck/R/x86_64-conda-linux-gnu-library/4.0</td><td>0.4.0 </td></tr>\n",
       "\t<tr><th scope=row>bitops</th><td>bitops  </td><td>/srv/home/cvandiedonck/R/x86_64-conda-linux-gnu-library/4.0</td><td>1.0-7 </td></tr>\n",
       "\t<tr><th scope=row>boot</th><td>boot    </td><td>/srv/home/cvandiedonck/R/x86_64-conda-linux-gnu-library/4.0</td><td>1.3-28</td></tr>\n",
       "\t<tr><th scope=row>cowplot</th><td>cowplot </td><td>/srv/home/cvandiedonck/R/x86_64-conda-linux-gnu-library/4.0</td><td>1.1.1 </td></tr>\n",
       "\t<tr><th scope=row>dabestr</th><td>dabestr </td><td>/srv/home/cvandiedonck/R/x86_64-conda-linux-gnu-library/4.0</td><td>0.3.0 </td></tr>\n",
       "\t<tr><th scope=row>effsize</th><td>effsize </td><td>/srv/home/cvandiedonck/R/x86_64-conda-linux-gnu-library/4.0</td><td>0.8.1 </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 6 √ó 3 of type chr\n",
       "\\begin{tabular}{r|lll}\n",
       "  & Package & LibPath & Version\\\\\n",
       "\\hline\n",
       "\tbeeswarm & beeswarm & /srv/home/cvandiedonck/R/x86\\_64-conda-linux-gnu-library/4.0 & 0.4.0 \\\\\n",
       "\tbitops & bitops   & /srv/home/cvandiedonck/R/x86\\_64-conda-linux-gnu-library/4.0 & 1.0-7 \\\\\n",
       "\tboot & boot     & /srv/home/cvandiedonck/R/x86\\_64-conda-linux-gnu-library/4.0 & 1.3-28\\\\\n",
       "\tcowplot & cowplot  & /srv/home/cvandiedonck/R/x86\\_64-conda-linux-gnu-library/4.0 & 1.1.1 \\\\\n",
       "\tdabestr & dabestr  & /srv/home/cvandiedonck/R/x86\\_64-conda-linux-gnu-library/4.0 & 0.3.0 \\\\\n",
       "\teffsize & effsize  & /srv/home/cvandiedonck/R/x86\\_64-conda-linux-gnu-library/4.0 & 0.8.1 \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 6 √ó 3 of type chr\n",
       "\n",
       "| <!--/--> | Package | LibPath | Version |\n",
       "|---|---|---|---|\n",
       "| beeswarm | beeswarm | /srv/home/cvandiedonck/R/x86_64-conda-linux-gnu-library/4.0 | 0.4.0  |\n",
       "| bitops | bitops   | /srv/home/cvandiedonck/R/x86_64-conda-linux-gnu-library/4.0 | 1.0-7  |\n",
       "| boot | boot     | /srv/home/cvandiedonck/R/x86_64-conda-linux-gnu-library/4.0 | 1.3-28 |\n",
       "| cowplot | cowplot  | /srv/home/cvandiedonck/R/x86_64-conda-linux-gnu-library/4.0 | 1.1.1  |\n",
       "| dabestr | dabestr  | /srv/home/cvandiedonck/R/x86_64-conda-linux-gnu-library/4.0 | 0.3.0  |\n",
       "| effsize | effsize  | /srv/home/cvandiedonck/R/x86_64-conda-linux-gnu-library/4.0 | 0.8.1  |\n",
       "\n"
      ],
      "text/plain": [
       "         Package  LibPath                                                    \n",
       "beeswarm beeswarm /srv/home/cvandiedonck/R/x86_64-conda-linux-gnu-library/4.0\n",
       "bitops   bitops   /srv/home/cvandiedonck/R/x86_64-conda-linux-gnu-library/4.0\n",
       "boot     boot     /srv/home/cvandiedonck/R/x86_64-conda-linux-gnu-library/4.0\n",
       "cowplot  cowplot  /srv/home/cvandiedonck/R/x86_64-conda-linux-gnu-library/4.0\n",
       "dabestr  dabestr  /srv/home/cvandiedonck/R/x86_64-conda-linux-gnu-library/4.0\n",
       "effsize  effsize  /srv/home/cvandiedonck/R/x86_64-conda-linux-gnu-library/4.0\n",
       "         Version\n",
       "beeswarm 0.4.0  \n",
       "bitops   1.0-7  \n",
       "boot     1.3-28 \n",
       "cowplot  1.1.1  \n",
       "dabestr  0.3.0  \n",
       "effsize  0.8.1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#cell 6\n",
    "head(installed.packages()[,c(1,2,3)])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Identifions √† pr√©sent quels paquets sont d√©j√† charg√©s dans notre session"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.0.5 (2021-03-31)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Ubuntu 18.04.4 LTS\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /srv/conda/envs/notebook/lib/libopenblasp-r0.3.17.so\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "attached base packages:\n",
       "[1] stats     graphics  grDevices utils     datasets  methods   base     \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       " [1] fansi_0.5.0       digest_0.6.27     utf8_1.2.2        crayon_1.4.1     \n",
       " [5] IRdisplay_1.0     repr_1.1.3        lifecycle_1.0.0   jsonlite_1.7.2   \n",
       " [9] evaluate_0.14     pillar_1.6.2      rlang_0.4.11      uuid_0.1-4       \n",
       "[13] vctrs_0.3.8       ellipsis_0.3.2    IRkernel_1.1.1    tools_4.0.5      \n",
       "[17] compiler_4.0.5    base64enc_0.1-3   pbdZMQ_0.3-5      htmltools_0.5.1.1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#cell 7\n",
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Installation et chargement de paquets\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "L‚Äôobjectif appliqu√© de ce tutoriel est d‚Äôutiliser le paquet `GenomicRanges` de ___Bioconductor___ afin de d√©tecter des chevauchements entre une liste de SNPs et des sites d‚ÄôH3K27ac (enhancers actifs). Nous utiliserons ensuite le paquet `biomaRt` √©galement de ___Bioconductor___ pour conna√Ætre les g√®nes les plus proches des chevauchements.\n",
    "\n",
    "Ces deux paquets appartenant au projet Bioconductor, la proc√©dure pour les installer repose sur la fonction `install()` du paquet `BiocManager` si vous avez une version de R >=3.5.\n",
    "\n",
    "Voyons comment installer et/ou charger ces diff√©rents paquets.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\"><b>  Warning</b>. Dans beaucoup de tutoriels, vous trouvez encore l'installation de paquets de Bioconductor avec la fonction <b>biocLite()</b>. Cette m√©thode est obsol√®te pour les versions de R>=3.5.</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.1. Chargement/Installation d'un paquet classique"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "L'installation classique d'un paquet repose sur la fonction `install.packages()`. Le paquet `BiocManager()` de Bioconductor est le seul paquet de Bioconductor √† s'installer classiquement."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Avant de tenter de l'installer, nous pouvons v√©rifier que le paquet `BiocManager` est d√©j√† install√© sur adenine dans le r√©pertoire associ√© √† l'environnement."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".dl-inline {width: auto; margin:0; padding: 0}\n",
       ".dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}\n",
       ".dl-inline>dt::after {content: \":\\0020\"; padding-right: .5ex}\n",
       ".dl-inline>dt:not(:first-of-type) {padding-left: .5ex}\n",
       "</style><dl class=dl-inline><dt>Package</dt><dd>'BiocManager'</dd><dt>LibPath</dt><dd>'/srv/conda/envs/notebook/lib/R/library'</dd><dt>Version</dt><dd>'1.30.16'</dd></dl>\n"
      ],
      "text/latex": [
       "\\begin{description*}\n",
       "\\item[Package] 'BiocManager'\n",
       "\\item[LibPath] '/srv/conda/envs/notebook/lib/R/library'\n",
       "\\item[Version] '1.30.16'\n",
       "\\end{description*}\n"
      ],
      "text/markdown": [
       "Package\n",
       ":   'BiocManager'LibPath\n",
       ":   '/srv/conda/envs/notebook/lib/R/library'Version\n",
       ":   '1.30.16'\n",
       "\n"
      ],
      "text/plain": [
       "                                 Package \n",
       "                           \"BiocManager\" \n",
       "                                 LibPath \n",
       "\"/srv/conda/envs/notebook/lib/R/library\" \n",
       "                                 Version \n",
       "                               \"1.30.16\" "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#cell 8\n",
    "installed.packages()[\"BiocManager\",c(1,2,3)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Nous pouvons √† pr√©sent charger le paquet avec la fonction `library()`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Bioconductor version '3.12' is out-of-date; the current release version '3.13'\n",
      "  is available with R version '4.1'; see https://bioconductor.org/install\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#cell 9\n",
    "library(BiocManager)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Si le paquet n'avez pas encore √©t√© install√©, la commande ci-dessus aurait retourn√© une erreur. \n",
    "Dans ce cas, vous auriez pu utiliser la commande ci-dessous pour l'installer avec la fonction `install.packages()`."
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "#raw cell\n",
    "install.packages(\"BiocManager\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\"><b> Warning</b>.Cependant, il est inutile de r√©installer un paquet d√©j√† install√©.\n",
    "\n",
    "L'execution conditionnelle ci-dessous solutionne le probl√®me, que le paquet soit d√©j√† install√© ou pas. En effet, cette commande permet de n'installer le paquet que s'il ne l'est pas encore. C'est ainsi que nous vous conseillons d'invoquer toutes les installations de paquets dans vos scripts afin de permettre une portabilit√© entre utilisateurs, en leur √©pargnant la r√©installation de paquets qu'ils ont d√©j√†. *(Nous verrons les √©x√©cutions conditionnelles plus en d√©tail lors de la session 6 de R.)*\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#cell 10\n",
    "if (!requireNamespace(\"BiocManager\"))\n",
    "  install.packages(\"BiocManager\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "V√©rifions √† pr√©sent que le paquet BiocManager a bien √©t√© charg√© et identifions sa version:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.0.5 (2021-03-31)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Ubuntu 18.04.4 LTS\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /srv/conda/envs/notebook/lib/libopenblasp-r0.3.17.so\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "attached base packages:\n",
       "[1] stats     graphics  grDevices utils     datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       "[1] BiocManager_1.30.16\n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       " [1] fansi_0.5.0       digest_0.6.27     utf8_1.2.2        crayon_1.4.1     \n",
       " [5] IRdisplay_1.0     repr_1.1.3        lifecycle_1.0.0   jsonlite_1.7.2   \n",
       " [9] evaluate_0.14     pillar_1.6.2      rlang_0.4.11      uuid_0.1-4       \n",
       "[13] vctrs_0.3.8       ellipsis_0.3.2    IRkernel_1.1.1    tools_4.0.5      \n",
       "[17] compiler_4.0.5    base64enc_0.1-3   pbdZMQ_0.3-5      htmltools_0.5.1.1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#cell 11\n",
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.2. Chargement/Installation d'un paquet de Bioconductor"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Nous continuons √† pr√©sent avec les deux paquets de Bioconductor dont nous aurons besoin pour notre analyse."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\"><b> Warning</b>. Comme indiqu√© pr√©c√©demment, les paquets de bioconductor ne s'installent pas avec la fonction install.packages() mais <b>avec la fonction intall() de BiocManager depuis R.3.5 </b>. </div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2.2.1. Paquet non install√©"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Le paquet `GenomicRanges` n'est pas encore install√© comme l'indique l'erreur de la commande ci-dessous."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in installed.packages()[\"GenomicRanges\", c(1, 2, 3)]: subscript out of bounds\n",
     "output_type": "error",
     "traceback": [
      "Error in installed.packages()[\"GenomicRanges\", c(1, 2, 3)]: subscript out of bounds\nTraceback:\n"
     ]
    }
   ],
   "source": [
    "#cell 12\n",
    "installed.packages()[\"GenomicRanges\",c(1,2,3)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Le paquet `BiocManager` √©tant bien install√©, nous pouvons installer `GenomicRanges`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required namespace: GenomicRanges\n",
      "\n",
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.12 (BiocManager 1.30.16), R 4.0.5 (2021-03-31)\n",
      "\n",
      "Installing package(s) 'GenomicRanges'\n",
      "\n",
      "Installation paths not writeable, unable to update packages\n",
      "  path: /srv/conda/envs/notebook/lib/R/library\n",
      "  packages:\n",
      "    data.table, dbplyr, desc, digest, DT, gert, hms, htmltools, IRkernel,\n",
      "    knitr, later, lattice, lifecycle, lubridate, matrixStats, memoise, mgcv,\n",
      "    mime, pbkrtest, pillar, rcmdcheck, RcppArmadillo, readr, remotes,\n",
      "    rmarkdown, roxygen2, stringi, testthat, tibble, tidyr, tidyverse, tinytex,\n",
      "    xfun, XML\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Loading required package: parallel\n",
      "\n",
      "\n",
      "Attaching package: ‚ÄòBiocGenerics‚Äô\n",
      "\n",
      "\n",
      "The following objects are masked from ‚Äòpackage:parallel‚Äô:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "The following objects are masked from ‚Äòpackage:stats‚Äô:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‚Äòpackage:base‚Äô:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‚ÄòS4Vectors‚Äô\n",
      "\n",
      "\n",
      "The following object is masked from ‚Äòpackage:base‚Äô:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#cell 13\n",
    "if (!requireNamespace(\"GenomicRanges\"))\n",
    "  BiocManager::install(\"GenomicRanges\")\n",
    "library(GenomicRanges)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Vous pouvez v√©rifier qu'il a bien √©t√© install√© dans votre r√©pertoire personnel:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".dl-inline {width: auto; margin:0; padding: 0}\n",
       ".dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}\n",
       ".dl-inline>dt::after {content: \":\\0020\"; padding-right: .5ex}\n",
       ".dl-inline>dt:not(:first-of-type) {padding-left: .5ex}\n",
       "</style><dl class=dl-inline><dt>Package</dt><dd>'GenomicRanges'</dd><dt>LibPath</dt><dd>'/srv/home/cvandiedonck/R/x86_64-conda-linux-gnu-library/4.0'</dd><dt>Version</dt><dd>'1.42.0'</dd></dl>\n"
      ],
      "text/latex": [
       "\\begin{description*}\n",
       "\\item[Package] 'GenomicRanges'\n",
       "\\item[LibPath] '/srv/home/cvandiedonck/R/x86\\_64-conda-linux-gnu-library/4.0'\n",
       "\\item[Version] '1.42.0'\n",
       "\\end{description*}\n"
      ],
      "text/markdown": [
       "Package\n",
       ":   'GenomicRanges'LibPath\n",
       ":   '/srv/home/cvandiedonck/R/x86_64-conda-linux-gnu-library/4.0'Version\n",
       ":   '1.42.0'\n",
       "\n"
      ],
      "text/plain": [
       "                                                      Package \n",
       "                                              \"GenomicRanges\" \n",
       "                                                      LibPath \n",
       "\"/srv/home/cvandiedonck/R/x86_64-conda-linux-gnu-library/4.0\" \n",
       "                                                      Version \n",
       "                                                     \"1.42.0\" "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#cell 14\n",
    "installed.packages()[\"GenomicRanges\",c(1,2,3)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.0.5 (2021-03-31)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Ubuntu 18.04.4 LTS\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /srv/conda/envs/notebook/lib/libopenblasp-r0.3.17.so\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "attached base packages:\n",
       "[1] parallel  stats4    stats     graphics  grDevices utils     datasets \n",
       "[8] methods   base     \n",
       "\n",
       "other attached packages:\n",
       "[1] GenomicRanges_1.42.0 GenomeInfoDb_1.26.7  IRanges_2.24.1      \n",
       "[4] S4Vectors_0.28.1     BiocGenerics_0.36.1  BiocManager_1.30.16 \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       " [1] XVector_0.30.0         zlibbioc_1.36.0        uuid_0.1-4            \n",
       " [4] rlang_0.4.11           fansi_0.5.0            tools_4.0.5           \n",
       " [7] utf8_1.2.2             htmltools_0.5.1.1      ellipsis_0.3.2        \n",
       "[10] digest_0.6.27          lifecycle_1.0.0        crayon_1.4.1          \n",
       "[13] GenomeInfoDbData_1.2.4 IRdisplay_1.0          repr_1.1.3            \n",
       "[16] base64enc_0.1-3        vctrs_0.3.8            bitops_1.0-7          \n",
       "[19] RCurl_1.98-1.5         IRkernel_1.1.1         evaluate_0.14         \n",
       "[22] pbdZMQ_0.3-5           compiler_4.0.5         pillar_1.6.2          \n",
       "[25] jsonlite_1.7.2        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#cell 15\n",
    "sessionInfo()# vous noterez que l'installation de GenomicRanges a n√©cessite celle de IRanges"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2.2.2. Paquet d√©j√† install√©"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".dl-inline {width: auto; margin:0; padding: 0}\n",
       ".dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}\n",
       ".dl-inline>dt::after {content: \":\\0020\"; padding-right: .5ex}\n",
       ".dl-inline>dt:not(:first-of-type) {padding-left: .5ex}\n",
       "</style><dl class=dl-inline><dt>Package</dt><dd>'biomaRt'</dd><dt>LibPath</dt><dd>'/srv/conda/envs/notebook/lib/R/library'</dd><dt>Version</dt><dd>'2.46.3'</dd></dl>\n"
      ],
      "text/latex": [
       "\\begin{description*}\n",
       "\\item[Package] 'biomaRt'\n",
       "\\item[LibPath] '/srv/conda/envs/notebook/lib/R/library'\n",
       "\\item[Version] '2.46.3'\n",
       "\\end{description*}\n"
      ],
      "text/markdown": [
       "Package\n",
       ":   'biomaRt'LibPath\n",
       ":   '/srv/conda/envs/notebook/lib/R/library'Version\n",
       ":   '2.46.3'\n",
       "\n"
      ],
      "text/plain": [
       "                                 Package \n",
       "                               \"biomaRt\" \n",
       "                                 LibPath \n",
       "\"/srv/conda/envs/notebook/lib/R/library\" \n",
       "                                 Version \n",
       "                                \"2.46.3\" "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#cell 16\n",
    "installed.packages()[\"biomaRt\",c(1,2,3)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "En r√©alit√©, que le paquet ait d√©j√† √©t√© install√© ou non, la commande avec l'√©x√©cution conditionnelle est la m√™me. <mark>Vous n'avez donc pas besoin de v√©rifier l'installation auparavant.</mark>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required namespace: biomaRt\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#cell 17\n",
    "# installation et/ou chargement de biomaRt:\n",
    "if (!requireNamespace(\"biomaRt\"))\n",
    "  BiocManager::install(\"biomaRt\")\n",
    "library(biomaRt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.0.5 (2021-03-31)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Ubuntu 18.04.4 LTS\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /srv/conda/envs/notebook/lib/libopenblasp-r0.3.17.so\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "attached base packages:\n",
       "[1] parallel  stats4    stats     graphics  grDevices utils     datasets \n",
       "[8] methods   base     \n",
       "\n",
       "other attached packages:\n",
       "[1] biomaRt_2.46.3       GenomicRanges_1.42.0 GenomeInfoDb_1.26.7 \n",
       "[4] IRanges_2.24.1       S4Vectors_0.28.1     BiocGenerics_0.36.1 \n",
       "[7] BiocManager_1.30.16 \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       " [1] pbdZMQ_0.3-5           progress_1.2.2         tidyselect_1.1.1      \n",
       " [4] repr_1.1.3             purrr_0.3.4            vctrs_0.3.8           \n",
       " [7] generics_0.1.0         htmltools_0.5.1.1      BiocFileCache_1.14.0  \n",
       "[10] base64enc_0.1-3        utf8_1.2.2             blob_1.2.2            \n",
       "[13] XML_3.99-0.7           rlang_0.4.11           pillar_1.6.2          \n",
       "[16] glue_1.4.2             DBI_1.1.1              rappdirs_0.3.3        \n",
       "[19] bit64_4.0.5            dbplyr_1.4.3           uuid_0.1-4            \n",
       "[22] GenomeInfoDbData_1.2.4 lifecycle_1.0.0        stringr_1.4.0         \n",
       "[25] zlibbioc_1.36.0        evaluate_0.14          memoise_1.1.0         \n",
       "[28] Biobase_2.50.0         curl_4.3.2             AnnotationDbi_1.52.0  \n",
       "[31] fansi_0.5.0            IRdisplay_1.0          Rcpp_1.0.7            \n",
       "[34] openssl_1.4.5          IRkernel_1.1.1         jsonlite_1.7.2        \n",
       "[37] XVector_0.30.0         bit_4.0.4              hms_1.1.0             \n",
       "[40] askpass_1.1            digest_0.6.27          stringi_1.7.4         \n",
       "[43] dplyr_1.0.7            tools_4.0.5            bitops_1.0-7          \n",
       "[46] magrittr_2.0.1         RCurl_1.98-1.5         RSQLite_2.2.8         \n",
       "[49] tibble_3.1.4           crayon_1.4.1           pkgconfig_2.0.3       \n",
       "[52] ellipsis_0.3.2         xml2_1.3.2             prettyunits_1.1.1     \n",
       "[55] assertthat_0.2.1       httr_1.4.2             R6_2.5.1              \n",
       "[58] compiler_4.0.5        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#cell 18\n",
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Utilisation de paquets: application √† des donn√©es g√©nomiques"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.1. Importez les data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Importez les donn√©es de SNPs et de sites d‚Äôac√©tylation (fichiers sur moodle ou dans `/srv/data/meg-m1-ue5/`)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "'data.frame':\t62 obs. of  26 variables:\n",
      " $ bin            : int  913 913 913 913 913 913 913 913 913 913 ...\n",
      " $ chrom          : chr  \"chr17\" \"chr17\" \"chr17\" \"chr17\" ...\n",
      " $ chromStart     : int  43024272 43024274 43024279 43024293 43024296 43024629 43024670 43024813 43025042 43025790 ...\n",
      " $ chromEnd       : int  43024273 43024275 43024280 43024294 43024297 43024630 43024671 43024814 43025043 43025791 ...\n",
      " $ name           : chr  \"rs7224581\" \"rs71367982\" \"rs7222911\" \"rs7224048\" ...\n",
      " $ score          : int  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ strand         : chr  \"+\" \"+\" \"+\" \"+\" ...\n",
      " $ refNCBI        : chr  \"A\" \"G\" \"G\" \"C\" ...\n",
      " $ refUCSC        : chr  \"A\" \"G\" \"G\" \"C\" ...\n",
      " $ observed       : chr  \"A/C\" \"C/G\" \"A/C/G\" \"A/C\" ...\n",
      " $ molType        : chr  \"genomic\" \"genomic\" \"genomic\" \"genomic\" ...\n",
      " $ class          : chr  \"single\" \"single\" \"single\" \"single\" ...\n",
      " $ valid          : chr  \"by-cluster,by-frequency,by-2hit-2allele,by-hapmap,by-1000genomes\" \"by-cluster,by-frequency,by-1000genomes\" \"by-cluster,by-frequency,by-2hit-2allele,by-hapmap\" \"by-cluster,by-frequency,by-2hit-2allele\" ...\n",
      " $ avHet          : num  0.479 0.479 0.5 0.5 0.48 ...\n",
      " $ avHetSE        : num  0.1 0.1 0 0 0.098 ...\n",
      " $ func           : chr  \"near-gene-5\" \"near-gene-5\" \"near-gene-5\" \"near-gene-5\" ...\n",
      " $ locType        : chr  \"exact\" \"exact\" \"exact\" \"exact\" ...\n",
      " $ weight         : int  1 1 1 1 1 1 1 1 1 1 ...\n",
      " $ exceptions     : chr  \"MixedObserved\" \"10\" \"SingleClassTriAllelic,InconsistentAlleles\" \"15\" ...\n",
      " $ submitterCount : chr  \"16\" \"1000GENOMES,BL,CSHL,DDI,GNOMAD,GRF,HUMANGENOME_JCVI,SWEGEN,TOPMED,USC_VALOUEV,\" \"17\" \"ABI,BCM_SSAHASNP,COMPLETE_GENOMICS,CSHL,DDI,ENSEMBL,GNOMAD,GRF,HAMMER_LAB,HUMANGENOME_JCVI,SSAHASNP,SWEGEN,SYST\"| __truncated__ ...\n",
      " $ submitters     : chr  \"1000GENOMES,BCM_SSAHASNP,BL,CSHL,CSHL-HAPMAP,DDI,ENSEMBL,GNOMAD,GRF,HUMANGENOME_JCVI,SC_SNP,SSAHASNP,SWEGEN,TOP\"| __truncated__ \"2\" \"BCM_SSAHASNP,BL,COMPLETE_GENOMICS,CSHL,CSHL-HAPMAP,DDI,ENSEMBL,GNOMAD,GRF,HAMMER_LAB,HUMANGENOME_JCVI,SC_SNP,SS\"| __truncated__ \"2\" ...\n",
      " $ alleleFreqCount: chr  \"2\" \"C,G,\" \"2\" \"A,C,\" ...\n",
      " $ alleles        : chr  \"A,C,\" \"3018.000000,1990.000000,\" \"A,G,\" \"6.000000,6.000000,\" ...\n",
      " $ alleleNs       : chr  \"1990.000000,3018.000000,\" \"0.602636,0.397364,\" \"8.000000,8.000000,\" \"0.500000,0.500000,\" ...\n",
      " $ alleleFreqs    : chr  \"0.397364,0.602636,\" \"maf-5-some-pop,maf-5-all-pops\" \"0.500000,0.500000,\" \"\" ...\n",
      " $ bitfields      : chr  \"maf-5-some-pop,maf-5-all-pops\" \"\" \"\" \"\" ...\n"
     ]
    }
   ],
   "source": [
    "#cell 19\n",
    "snps <- read.table(\"/srv/data/meg-m1-ue5/hgTables_SNPs.txt\", header = T, stringsAsFactors = F, fill = T)\n",
    "str(snps)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "'data.frame':\t30 obs. of  4 variables:\n",
      " $ V1: chr  \"chr17\" \"chr17\" \"chr17\" \"chr17\" ...\n",
      " $ V2: int  43023996 43024283 43024658 43025608 43026183 43027058 43027833 43028208 43029008 43029258 ...\n",
      " $ V3: int  43024058 43024633 43025383 43025833 43026458 43027258 43028183 43028408 43029233 43029508 ...\n",
      " $ V4: chr  \"chr17.1\" \"chr17.2\" \"chr17.3\" \"chr17.4\" ...\n"
     ]
    }
   ],
   "source": [
    "#cell 20\n",
    "h3k27ac <- read.table(\"/srv/data/meg-m1-ue5/hgTables_H3K27Ac.txt\", stringsAsFactors = F, skip = 1)\n",
    "str(h3k27ac)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Donnez un nom de colonnes au dataframe des cites d‚Äôac√©tylation (¬´ chrom ¬ª, ¬´ start ¬ª, ¬´ end ¬ª, ¬´ id ¬ª)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "'data.frame':\t30 obs. of  4 variables:\n",
      " $ chrom: chr  \"chr17\" \"chr17\" \"chr17\" \"chr17\" ...\n",
      " $ start: int  43023996 43024283 43024658 43025608 43026183 43027058 43027833 43028208 43029008 43029258 ...\n",
      " $ end  : int  43024058 43024633 43025383 43025833 43026458 43027258 43028183 43028408 43029233 43029508 ...\n",
      " $ id   : chr  \"chr17.1\" \"chr17.2\" \"chr17.3\" \"chr17.4\" ...\n"
     ]
    }
   ],
   "source": [
    "#cell 21\n",
    "names(h3k27ac) <- c(\"chrom\", \"start\", \"end\", \"id\")\n",
    "str(h3k27ac)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.2. Utilisation du paquet **GenomicRanges**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3.2.1. Cherchez de la documentation sur ce paquet\n",
    "\n",
    "- soit √† https://www.bioconductor.org/packages/release/bioc/html/GenomicRanges.html\n",
    "- ou en consultant la vignette:\n",
    "`browseVignettes(\"GenomicRanges\")`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\"><b> Warning</b>. Le chargement de vignettes ne fonctionne pas dans les notebooks jupyter.</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3.2.2. Utilisation de la fonction `makeGRangesFromDataFrame` du paquet **GenomicRanges**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Cette fonction va g√©n√©rer un objet de type GRanges. Regardez la documentation correspondante."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table width=\"100%\" summary=\"page for GRanges-class {GenomicRanges}\"><tr><td>GRanges-class {GenomicRanges}</td><td style=\"text-align: right;\">R Documentation</td></tr></table>\n",
       "\n",
       "<h2>GRanges objects</h2>\n",
       "\n",
       "<h3>Description</h3>\n",
       "\n",
       "<p>The GRanges class is a container for the genomic locations and\n",
       "their associated annotations.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Details</h3>\n",
       "\n",
       "<p>GRanges is a vector of genomic locations and associated\n",
       "annotations. Each element in the vector is comprised of a sequence name,\n",
       "an interval, a strand, and optional metadata columns (e.g. score, GC\n",
       "content, etc.). This information is stored in four components:\n",
       "</p>\n",
       "\n",
       "<dl>\n",
       "<dt><code>seqnames</code></dt><dd><p>a 'factor' Rle object\n",
       "containing the sequence names.</p>\n",
       "</dd>\n",
       "<dt><code>ranges</code></dt><dd><p>an IRanges object containing\n",
       "the ranges.</p>\n",
       "</dd>\n",
       "<dt><code>strand</code></dt><dd><p>a 'factor' Rle object containing\n",
       "the strand information.</p>\n",
       "</dd>\n",
       "<dt><code>mcols</code></dt><dd><p>a DataFrame object\n",
       "containing the metadata columns. Columns cannot be named\n",
       "<code>\"seqnames\"</code>, <code>\"ranges\"</code>, <code>\"strand\"</code>,\n",
       "<code>\"seqlevels\"</code>, <code>\"seqlengths\"</code>, <code>\"isCircular\"</code>,\n",
       "<code>\"start\"</code>, <code>\"end\"</code>, <code>\"width\"</code>, or <code>\"element\"</code>.</p>\n",
       "</dd>\n",
       "<dt><code>seqinfo</code></dt><dd><p>a Seqinfo object containing information\n",
       "about the set of genomic sequences present in the GRanges object.</p>\n",
       "</dd>\n",
       "</dl>\n",
       "\n",
       "\n",
       "\n",
       "<h3>Constructor</h3>\n",
       "\n",
       "\n",
       "<dl>\n",
       "<dt></dt><dd>\n",
       "<p><code>GRanges(seqnames=NULL, ranges=NULL, strand=NULL,\n",
       "                    ..., seqinfo=NULL, seqlengths=NULL)</code>:\n",
       "Creates a GRanges object.\n",
       "</p>\n",
       "\n",
       "<dl>\n",
       "<dt><code>seqnames</code></dt><dd>\n",
       "<p><code>NULL</code>, or an Rle object, character vector,\n",
       "or factor containing the sequence names.\n",
       "</p>\n",
       "</dd>\n",
       "<dt><code>ranges</code></dt><dd>\n",
       "<p><code>NULL</code>, or an IRanges object containing the\n",
       "ranges.\n",
       "</p>\n",
       "</dd>\n",
       "<dt><code>strand</code></dt><dd>\n",
       "<p><code>NULL</code>, or an Rle object, character vector,\n",
       "or factor containing the strand information.\n",
       "</p>\n",
       "</dd>\n",
       "<dt><code>...</code></dt><dd>\n",
       "<p>Metadata columns to set on the GRanges object. All the metadata\n",
       "columns must be vector-like objects of the same length as the object\n",
       "to construct. They cannot be named <code>\"start\"</code>, <code>\"end\"</code>,\n",
       "<code>\"width\"</code>, or <code>\"element\"</code>.\n",
       "</p>\n",
       "</dd>\n",
       "<dt><code>seqinfo</code></dt><dd>\n",
       "<p>Either <code>NULL</code>, or a Seqinfo object,\n",
       "or a character vector of unique sequence names (a.k.a.\n",
       "<em>seqlevels</em>), or a named numeric vector of sequence lengths.\n",
       "When not <code>NULL</code>, <code>seqinfo</code> must be compatible with the\n",
       "sequence names in <code>seqnames</code>, that is, it must have one entry\n",
       "for each unique sequence name in <code>seqnames</code>. Note that it can\n",
       "have additional entries i.e. entries for seqlevels not present\n",
       "in <code>seqnames</code>.\n",
       "</p>\n",
       "</dd>\n",
       "<dt><code>seqlengths</code></dt><dd>\n",
       "<p><code>NULL</code>, or an integer vector named with <code>levels(seqnames)</code>\n",
       "and containing the lengths (or NA) for each level in\n",
       "<code>levels(seqnames)</code>.\n",
       "</p>\n",
       "</dd>\n",
       "</dl>\n",
       "\n",
       "<p>If <code>ranges</code> is not supplied and/or NULL then the constructor\n",
       "proceeds in 2 steps:\n",
       "</p>\n",
       "\n",
       "<ol>\n",
       "<li><p> An initial GRanges object is created with\n",
       "<code>as(seqnames, \"GRanges\")</code>.\n",
       "</p>\n",
       "</li>\n",
       "<li><p> Then this GRanges object is updated according to whatever\n",
       "non-NULL remaining arguments were passed to the call to\n",
       "<code>GRanges()</code>.\n",
       "</p>\n",
       "</li></ol>\n",
       "\n",
       "<p>As a consequence of this behavior, <code>GRanges(x)</code> is equivalent to\n",
       "<code>as(x, \"GRanges\")</code>.\n",
       "</p>\n",
       "</dd>\n",
       "</dl>\n",
       "\n",
       "\n",
       "\n",
       "<h3>Accessors</h3>\n",
       "\n",
       "<p>In the following code snippets, <code>x</code> is a GRanges object.\n",
       "</p>\n",
       "\n",
       "<dl>\n",
       "<dt></dt><dd>\n",
       "<p><code>length(x)</code>:\n",
       "Get the number of elements.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>seqnames(x)</code>, <code>seqnames(x) &lt;- value</code>:\n",
       "Get or set the sequence names.\n",
       "<code>value</code> can be an Rle object, a character vector,\n",
       "or a factor.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>ranges(x)</code>, <code>ranges(x) &lt;- value</code>:\n",
       "Get or set the ranges. <code>value</code> can be an\n",
       "IntegerRanges object.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>start(x)</code>, <code>start(x) &lt;- value</code>:\n",
       "Get or set <code>start(ranges(x))</code>.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>end(x)</code>, <code>end(x) &lt;- value</code>:\n",
       "Get or set <code>end(ranges(x))</code>.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>width(x)</code>, <code>width(x) &lt;- value</code>:\n",
       "Get or set <code>width(ranges(x))</code>.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>strand(x)</code>, <code>strand(x) &lt;- value</code>:\n",
       "Get or set the strand. <code>value</code> can be an Rle object, character\n",
       "vector, or factor.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>names(x)</code>, <code>names(x) &lt;- value</code>:\n",
       "Get or set the names of the elements.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>mcols(x, use.names=FALSE)</code>, <code>mcols(x) &lt;- value</code>:\n",
       "Get or set the metadata columns.\n",
       "If <code>use.names=TRUE</code> and the metadata columns are not <code>NULL</code>,\n",
       "then the names of <code>x</code> are propagated as the row names of the\n",
       "returned DataFrame object.\n",
       "When setting the metadata columns, the supplied value must be <code>NULL</code>\n",
       "or a data-frame-like object (i.e. DataFrame or data.frame)\n",
       "holding element-wise metadata.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>elementMetadata(x)</code>, <code>elementMetadata(x) &lt;- value</code>,\n",
       "<code>values(x)</code>, <code>values(x) &lt;- value</code>:\n",
       "Alternatives to <code>mcols</code> functions. Their use is discouraged.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>seqinfo(x)</code>, <code>seqinfo(x) &lt;- value</code>:\n",
       "Get or set the information about the underlying sequences.\n",
       "<code>value</code> must be a Seqinfo object.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>seqlevels(x)</code>,\n",
       "<code>seqlevels(x, pruning.mode=c(\"error\", \"coarse\", \"fine\", \"tidy\")) &lt;- value</code>:\n",
       "Get or set the sequence levels.\n",
       "<code>seqlevels(x)</code> is equivalent to <code>seqlevels(seqinfo(x))</code>\n",
       "or to <code>levels(seqnames(x))</code>, those 2 expressions being\n",
       "guaranteed to return identical character vectors on a GRanges object.\n",
       "<code>value</code> must be a character vector with no NAs.\n",
       "See <code>?seqlevels</code> for more information.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>seqlengths(x)</code>, <code>seqlengths(x) &lt;- value</code>:\n",
       "Get or set the sequence lengths.\n",
       "<code>seqlengths(x)</code> is equivalent to <code>seqlengths(seqinfo(x))</code>.\n",
       "<code>value</code> can be a named non-negative integer or numeric vector\n",
       "eventually with NAs.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>isCircular(x)</code>, <code>isCircular(x) &lt;- value</code>:\n",
       "Get or set the circularity flags.\n",
       "<code>isCircular(x)</code> is equivalent to <code>isCircular(seqinfo(x))</code>.\n",
       "<code>value</code> must be a named logical vector eventually with NAs.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>genome(x)</code>, <code>genome(x) &lt;- value</code>:\n",
       "Get or set the genome identifier or assembly name for each sequence.\n",
       "<code>genome(x)</code> is equivalent to <code>genome(seqinfo(x))</code>.\n",
       "<code>value</code> must be a named character vector eventually with NAs.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>seqlevelsStyle(x)</code>, <code>seqlevelsStyle(x) &lt;- value</code>:\n",
       "Get or set the seqname style for <code>x</code>.\n",
       "See the seqlevelsStyle generic getter and setter\n",
       "in the <span class=\"pkg\">GenomeInfoDb</span> package for more information.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>score(x), score(x) &lt;- value</code>: Get or set the &ldquo;score&rdquo;\n",
       "column from the element metadata.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>granges(x, use.names=FALSE, use.mcols=FALSE)</code>: Squeeze the genomic\n",
       "ranges out of GenomicRanges object <code>x</code> and return them in a\n",
       "GRanges object <em>parallel</em> to <code>x</code> (i.e. same length as <code>x</code>).\n",
       "If <code>use.mcols</code> is <code>TRUE</code>, the metadata columns are propagated.\n",
       "If <code>x</code> is a GenomicRanges derivative with <em>extra column\n",
       "slots</em>, these will be propagated as metadata columns on the returned\n",
       "GRanges object.\n",
       "</p>\n",
       "</dd>\n",
       "</dl>\n",
       "\n",
       "\n",
       "\n",
       "<h3>Coercion</h3>\n",
       "\n",
       "<p>In the code snippets below, <code>x</code> is a GRanges object.\n",
       "</p>\n",
       "\n",
       "<dl>\n",
       "<dt></dt><dd>\n",
       "<p><code>as(from, \"GRanges\")</code>: Creates a GRanges object from a character\n",
       "vector, a factor, or IntegerRangesList object.\n",
       "</p>\n",
       "<p>When <code>from</code> is a character vector (or a factor), each element\n",
       "in it must represent a genomic range in format <code>chr1:2501-2800</code>\n",
       "(unstranded range) or <code>chr1:2501-2800:+</code> (stranded range).\n",
       "<code>..</code> is also supported as a separator between the start and end\n",
       "positions. Strand can be <code>+</code>, <code>-</code>, <code>*</code>, or missing.\n",
       "The names on <code>from</code> are propagated to the returned GRanges object.\n",
       "See <code>as.character()</code> and <code>as.factor()</code> below for the\n",
       "reverse transformations.\n",
       "</p>\n",
       "<p>Coercing a data.frame or DataFrame into a GRanges object is also\n",
       "supported. See <code>makeGRangesFromDataFrame</code> for the details.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>as(from, \"IntegerRangesList\")</code>:\n",
       "Creates a IntegerRangesList object from a GRanges\n",
       "object. The <code>strand</code> and metadata columns become <em>inner</em>\n",
       "metadata columns (i.e. metadata columns on the ranges).\n",
       "The <code>seqlengths(from)</code>, <code>isCircular(from)</code>, and\n",
       "<code>genome(from)</code> vectors become the metadata columns.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>as.character(x, ignore.strand=FALSE)</code>:\n",
       "Turn GRanges object <code>x</code> into a character vector where each\n",
       "range in <code>x</code> is represented by a string in format\n",
       "<code>chr1:2501-2800:+</code>. If <code>ignore.strand</code> is TRUE or if\n",
       "<em>all</em> the ranges in <code>x</code> are unstranded (i.e. their strand\n",
       "is set to <code>*</code>), then all the strings in the output are in\n",
       "format <code>chr1:2501-2800</code>.\n",
       "</p>\n",
       "<p>The names on <code>x</code> are propagated to the returned character vector.\n",
       "Its metadata (<code>metadata(x)</code>) and metadata columns (<code>mcols(x)</code>)\n",
       "are ignored.\n",
       "</p>\n",
       "<p>See <code>as(from, \"GRanges\")</code> above for the reverse transformation.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>as.factor(x)</code>:\n",
       "Equivalent to\n",
       "</p>\n",
       "<pre>  factor(as.character(x), levels=as.character(sort(unique(x))))\n",
       "</pre>\n",
       "<p>See <code>as(from, \"GRanges\")</code> above for the reverse transformation.\n",
       "</p>\n",
       "<p>Note that <code>table(x)</code> is supported on a GRanges object. It is\n",
       "equivalent to, but much faster than, <code>table(as.factor(x))</code>.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>as.data.frame(x, row.names = NULL, optional = FALSE, ...)</code>:\n",
       "Creates a data.frame with columns <code>seqnames</code> (factor),\n",
       "<code>start</code> (integer), <code>end</code> (integer), <code>width</code> (integer),\n",
       "<code>strand</code> (factor), as well as the additional metadata columns\n",
       "stored in <code>mcols(x)</code>. Pass an explicit\n",
       "<code>stringsAsFactors=TRUE/FALSE</code> argument via <code>...</code> to\n",
       "override the default conversions for the metadata columns in\n",
       "<code>mcols(x)</code>.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>as(from, \"Grouping\")</code>: Creates a\n",
       "<code>ManyToOneGrouping</code> object that groups\n",
       "<code>from</code> by seqname, strand, start and end (same as the default\n",
       "sort order). This makes it convenient, for example, to aggregate a\n",
       "GenomicRanges object by range.\n",
       "</p>\n",
       "</dd>\n",
       "</dl>\n",
       "\n",
       "<p>In the code snippets below, <code>x</code> is a Seqinfo\n",
       "object.\n",
       "</p>\n",
       "\n",
       "<dl>\n",
       "<dt></dt><dd>\n",
       "<p><code>as(x, \"GRanges\")</code>, <code>as(x, \"GenomicRanges\")</code>,\n",
       "<code>as(x, \"IntegerRangesList\")</code>: Turns Seqinfo\n",
       "object <code>x</code> (with no <code>NA</code> lengths) into a GRanges or\n",
       "IntegerRangesList.\n",
       "</p>\n",
       "</dd>\n",
       "</dl>\n",
       "\n",
       "\n",
       "\n",
       "<h3>Subsetting</h3>\n",
       "\n",
       "<p>In the code snippets below, <code>x</code> is a GRanges object.\n",
       "</p>\n",
       "\n",
       "<dl>\n",
       "<dt></dt><dd>\n",
       "<p><code>x[i]</code>:\n",
       "Return a new GRanges object made of the elements selected by <code>i</code>.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>x[i, j]</code>:\n",
       "Like the above, but allow the user to conveniently subset the metadata\n",
       "columns thru <code>j</code>.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>x[i] &lt;- value</code>:\n",
       "Replacement version of <code>x[i]</code>.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>x$name</code>, <code>x$name &lt;- value</code>:\n",
       "Shortcuts for <code>mcols(x)$name</code> and <code>mcols(x)$name &lt;- value</code>,\n",
       "respectively. Provided as a convenience, for GRanges objects *only*,\n",
       "and as the result of strong popular demand.\n",
       "Note that those methods are not consistent with the other <code>$</code>\n",
       "and <code>$&lt;-</code> methods in the IRanges/GenomicRanges infrastructure,\n",
       "and might confuse some users by making them believe that a GRanges\n",
       "object can be manipulated as a data.frame-like object.\n",
       "Therefore we recommend using them only interactively, and we discourage\n",
       "their use in scripts or packages. For the latter, use\n",
       "<code>mcols(x)$name</code> and <code>mcols(x)$name &lt;- value</code>, instead\n",
       "of <code>x$name</code> and <code>x$name &lt;- value</code>, respectively.\n",
       "</p>\n",
       "</dd>\n",
       "</dl>\n",
       "\n",
       "<p>See <code>?`[`</code> in the <span class=\"pkg\">S4Vectors</span> package for more\n",
       "information about subsetting Vector derivatives and for an important note\n",
       "about the <code>x[i, j]</code> form.\n",
       "</p>\n",
       "<p>Note that a GRanges object can be used as a subscript to subset a\n",
       "list-like object that has names on it. In that case, the names on the\n",
       "list-like object are interpreted as sequence names.\n",
       "In the code snippets below, <code>x</code> is a list or List object with\n",
       "names on it, and the subscript <code>gr</code> is a GRanges object with all its\n",
       "seqnames being valid <code>x</code> names.\n",
       "</p>\n",
       "\n",
       "<dl>\n",
       "<dt></dt><dd>\n",
       "<p><code>x[gr]</code>:\n",
       "Return an object of the same class as <code>x</code> and <em>parallel</em>\n",
       "to <code>gr</code>. More precisely, it's conceptually doing:\n",
       "</p>\n",
       "<pre>  lapply(gr, function(gr1) x[[seqnames(gr1)]][ranges(gr1)])\n",
       "</pre>\n",
       "</dd>\n",
       "</dl>\n",
       "\n",
       "\n",
       "\n",
       "<h3>Concatenation</h3>\n",
       "\n",
       "\n",
       "<dl>\n",
       "<dt></dt><dd>\n",
       "<p><code>c(x, ..., ignore.mcols=FALSE)</code>:\n",
       "Concatenate GRanges object <code>x</code> and the GRanges objects in\n",
       "<code>...</code> together.\n",
       "See <code>?c</code> in the <span class=\"pkg\">S4Vectors</span> package\n",
       "for more information about concatenating Vector derivatives.\n",
       "</p>\n",
       "</dd>\n",
       "</dl>\n",
       "\n",
       "\n",
       "\n",
       "<h3>Splitting</h3>\n",
       "\n",
       "\n",
       "<dl>\n",
       "<dt></dt><dd>\n",
       "<p><code>split(x, f, drop=FALSE)</code>:\n",
       "Splits GRanges object <code>x</code> according to <code>f</code> to create a\n",
       "GRangesList object.\n",
       "If <code>f</code> is a list-like object then <code>drop</code> is ignored\n",
       "and <code>f</code> is treated as if it was\n",
       "<code>rep(seq_len(length(f)), sapply(f, length))</code>,\n",
       "so the returned object has the same shape as <code>f</code> (it also\n",
       "receives the names of <code>f</code>).\n",
       "Otherwise, if <code>f</code> is not a list-like object, empty list\n",
       "elements are removed from the returned object if <code>drop</code> is\n",
       "<code>TRUE</code>.\n",
       "</p>\n",
       "</dd>\n",
       "</dl>\n",
       "\n",
       "\n",
       "\n",
       "<h3>Displaying</h3>\n",
       "\n",
       "<p>In the code snippets below, <code>x</code> is a GRanges object.\n",
       "</p>\n",
       "\n",
       "<dl>\n",
       "<dt></dt><dd>\n",
       "<p><code>show(x)</code>:\n",
       "By default the <code>show</code> method displays 5 head and 5 tail\n",
       "elements. This can be changed by setting the global options\n",
       "<code>showHeadLines</code> and <code>showTailLines</code>. If the object\n",
       "length is less than (or equal to) the sum of these 2 options\n",
       "plus 1, then the full object is displayed.\n",
       "Note that these options also affect the display of\n",
       "GAlignments and\n",
       "GAlignmentPairs objects (defined in\n",
       "the <span class=\"pkg\">GenomicAlignments</span> package), as well as other objects\n",
       "defined in the <span class=\"pkg\">IRanges</span> and <span class=\"pkg\">Biostrings</span> packages (e.g.\n",
       "IRanges and DNAStringSet objects).\n",
       "</p>\n",
       "</dd>\n",
       "</dl>\n",
       "\n",
       "\n",
       "\n",
       "<h3>Author(s)</h3>\n",
       "\n",
       "<p>P. Aboyoun and H. Pag√®s</p>\n",
       "\n",
       "\n",
       "<h3>See Also</h3>\n",
       "\n",
       "\n",
       "<ul>\n",
       "<li><p> The IRanges class in the <span class=\"pkg\">IRanges</span>\n",
       "package for storing a set of <em>integer ranges</em>.\n",
       "</p>\n",
       "</li>\n",
       "<li><p> The GPos class for representing a set of <em>genomic\n",
       "positions</em> (i.e. <em>genomic ranges</em> of width 1, a.k.a.\n",
       "<em>genomic loci</em>).\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>makeGRangesFromDataFrame</code> for making a GRanges object\n",
       "from a data.frame or DataFrame object.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p>Seqinfo objects and the\n",
       "<code>seqinfo</code> accessor and family in the\n",
       "<span class=\"pkg\">GenomeInfoDb</span> package for accessing/modifying information\n",
       "about the underlying sequences of a GenomicRanges derivative.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p>GenomicRanges-comparison for comparing and ordering genomic\n",
       "ranges and/or positions.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p>findOverlaps-methods for finding overlapping\n",
       "genomic ranges and/or positions.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p>intra-range-methods and\n",
       "inter-range-methods\n",
       "for <em>intra range</em> and <em>inter range</em> transformations\n",
       "of GenomicRanges derivatives.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p>coverage-methods for computing the coverage\n",
       "of a set of genomic ranges and/or positions.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p>setops-methods for set operations on GRanges\n",
       "objects.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p>nearest-methods for finding the nearest\n",
       "genomic range/position neighbor.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>absoluteRanges</code> for transforming genomic ranges into\n",
       "<em>absolute</em> ranges (i.e. into ranges on the sequence obtained\n",
       "by virtually concatenating all the sequences in a genome).\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>tileGenome</code> for putting tiles on a genome.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p>genomicvars for manipulating genomic variables.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p>GRangesList objects.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p>Vector, Rle, and\n",
       "DataFrame objects in the <span class=\"pkg\">S4Vectors</span> package.\n",
       "</p>\n",
       "</li></ul>\n",
       "\n",
       "\n",
       "\n",
       "<h3>Examples</h3>\n",
       "\n",
       "<pre>\n",
       "showClass(\"GRanges\")  # shows the known subclasses\n",
       "\n",
       "## ---------------------------------------------------------------------\n",
       "## CONSTRUCTION\n",
       "## ---------------------------------------------------------------------\n",
       "## Specifying the bare minimum i.e. seqnames and ranges only. The\n",
       "## GRanges object will have no names, no strand information, and no\n",
       "## metadata columns:\n",
       "gr0 &lt;- GRanges(Rle(c(\"chr2\", \"chr2\", \"chr1\", \"chr3\"), c(1, 3, 2, 4)),\n",
       "               IRanges(1:10, width=10:1))\n",
       "gr0\n",
       "\n",
       "## Specifying names, strand, metadata columns. They can be set on an\n",
       "## existing object:\n",
       "names(gr0) &lt;- head(letters, 10)\n",
       "strand(gr0) &lt;- Rle(strand(c(\"-\", \"+\", \"*\", \"+\", \"-\")), c(1, 2, 2, 3, 2))\n",
       "mcols(gr0)$score &lt;- 1:10\n",
       "mcols(gr0)$GC &lt;- seq(1, 0, length=10)\n",
       "gr0\n",
       "\n",
       "## ... or specified at construction time:\n",
       "gr &lt;- GRanges(Rle(c(\"chr2\", \"chr2\", \"chr1\", \"chr3\"), c(1, 3, 2, 4)),\n",
       "              IRanges(1:10, width=10:1, names=head(letters, 10)),\n",
       "              Rle(strand(c(\"-\", \"+\", \"*\", \"+\", \"-\")), c(1, 2, 2, 3, 2)),\n",
       "              score=1:10, GC=seq(1, 0, length=10))\n",
       "stopifnot(identical(gr0, gr))\n",
       "\n",
       "## Specifying the seqinfo. It can be set on an existing object:\n",
       "seqinfo &lt;- Seqinfo(paste0(\"chr\", 1:3), c(1000, 2000, 1500), NA, \"mock1\")\n",
       "seqinfo(gr0) &lt;- merge(seqinfo(gr0), seqinfo)\n",
       "seqlevels(gr0) &lt;- seqlevels(seqinfo)\n",
       "\n",
       "## ... or specified at construction time:\n",
       "gr &lt;- GRanges(Rle(c(\"chr2\", \"chr2\", \"chr1\", \"chr3\"), c(1, 3, 2, 4)),\n",
       "              IRanges(1:10, width=10:1, names=head(letters, 10)),\n",
       "              Rle(strand(c(\"-\", \"+\", \"*\", \"+\", \"-\")), c(1, 2, 2, 3, 2)),\n",
       "              score=1:10, GC=seq(1, 0, length=10),\n",
       "              seqinfo=seqinfo)\n",
       "stopifnot(identical(gr0, gr))\n",
       "\n",
       "## ---------------------------------------------------------------------\n",
       "## COERCION\n",
       "## ---------------------------------------------------------------------\n",
       "## From GRanges:\n",
       "as.character(gr)\n",
       "as.factor(gr)\n",
       "as.data.frame(gr)\n",
       "\n",
       "## From character to GRanges:\n",
       "x1 &lt;- \"chr2:56-125\"\n",
       "as(x1, \"GRanges\")\n",
       "as(rep(x1, 4), \"GRanges\")\n",
       "x2 &lt;- c(A=x1, B=\"chr1:25-30:-\")\n",
       "as(x2, \"GRanges\")\n",
       "\n",
       "## From data.frame to GRanges:\n",
       "df &lt;- data.frame(chrom=\"chr2\", start=11:15, end=20:24)\n",
       "gr3 &lt;- as(df, \"GRanges\")\n",
       "\n",
       "## Alternatively, coercion to GRanges can be done by just calling the\n",
       "## GRanges() constructor on the object to coerce:\n",
       "gr1 &lt;- GRanges(x1)  # same as as(x1, \"GRanges\")\n",
       "gr2 &lt;- GRanges(x2)  # same as as(x2, \"GRanges\")\n",
       "gr3 &lt;- GRanges(df)  # same as as(df, \"GRanges\")\n",
       "\n",
       "## Sanity checks:\n",
       "stopifnot(identical(as(x1, \"GRanges\"), gr1))\n",
       "stopifnot(identical(as(x2, \"GRanges\"), gr2))\n",
       "stopifnot(identical(as(df, \"GRanges\"), gr3))\n",
       "\n",
       "## ---------------------------------------------------------------------\n",
       "## SUMMARIZING ELEMENTS\n",
       "## ---------------------------------------------------------------------\n",
       "table(seqnames(gr))\n",
       "table(strand(gr))\n",
       "sum(width(gr))\n",
       "table(gr)\n",
       "summary(mcols(gr)[,\"score\"])\n",
       "\n",
       "## The number of lines displayed in the 'show' method are controlled\n",
       "## with two global options:\n",
       "longGR  &lt;- sample(gr, 25, replace=TRUE)\n",
       "longGR\n",
       "options(showHeadLines=7)\n",
       "options(showTailLines=2)\n",
       "longGR\n",
       "\n",
       "## Revert to default values\n",
       "options(showHeadLines=NULL)\n",
       "options(showTailLines=NULL)\n",
       "\n",
       "## ---------------------------------------------------------------------\n",
       "## INVERTING THE STRAND\n",
       "## ---------------------------------------------------------------------\n",
       "invertStrand(gr)\n",
       "\n",
       "## ---------------------------------------------------------------------\n",
       "## RENAMING THE UNDERLYING SEQUENCES\n",
       "## ---------------------------------------------------------------------\n",
       "seqlevels(gr)\n",
       "seqlevels(gr) &lt;- sub(\"chr\", \"Chrom\", seqlevels(gr))\n",
       "gr\n",
       "seqlevels(gr) &lt;- sub(\"Chrom\", \"chr\", seqlevels(gr)) # revert\n",
       "\n",
       "## ---------------------------------------------------------------------\n",
       "## COMBINING OBJECTS\n",
       "## ---------------------------------------------------------------------\n",
       "gr2 &lt;- GRanges(seqnames=Rle(c('chr1', 'chr2', 'chr3'), c(3, 3, 4)),\n",
       "               IRanges(1:10, width=5),\n",
       "               strand='-',\n",
       "               score=101:110, GC=runif(10),\n",
       "               seqinfo=seqinfo)\n",
       "gr3 &lt;- GRanges(seqnames=Rle(c('chr1', 'chr2', 'chr3'), c(3, 4, 3)),\n",
       "               IRanges(101:110, width=10),\n",
       "               strand='-',\n",
       "               score=21:30,\n",
       "               seqinfo=seqinfo)\n",
       "some.gr &lt;- c(gr, gr2)\n",
       "\n",
       "c(gr, gr2, gr3)\n",
       "c(gr, gr2, gr3, ignore.mcols=TRUE)\n",
       "\n",
       "## ---------------------------------------------------------------------\n",
       "## USING A GRANGES OBJECT AS A SUBSCRIPT TO SUBSET ANOTHER OBJECT\n",
       "## ---------------------------------------------------------------------\n",
       "## Subsetting *by* a GRanges subscript is supported only if the object\n",
       "## to subset is a named list-like object:\n",
       "x &lt;- RleList(chr1=101:120, chr2=2:-8, chr3=31:40)\n",
       "x[gr]\n",
       "</pre>\n",
       "\n",
       "<hr /><div style=\"text-align: center;\">[Package <em>GenomicRanges</em> version 1.42.0 ]</div>"
      ],
      "text/latex": [
       "\\inputencoding{utf8}\n",
       "\\HeaderA{GRanges-class}{GRanges objects}{GRanges.Rdash.class}\n",
       "\\aliasA{[,list\\_OR\\_List,GenomicRanges-method}{GRanges-class}{[,list.Rul.OR.Rul.List,GenomicRanges.Rdash.method}\n",
       "\\aliasA{\\$,GenomicRanges-method}{GRanges-class}{.Rdol.,GenomicRanges.Rdash.method}\n",
       "\\aliasA{\\$<\\Rdash{},GenomicRanges-method}{GRanges-class}{.Rdol.<.Rdash.,GenomicRanges.Rdash.method}\n",
       "\\aliasA{as.character,GenomicRanges-method}{GRanges-class}{as.character,GenomicRanges.Rdash.method}\n",
       "\\aliasA{as.data.frame,GenomicRanges-method}{GRanges-class}{as.data.frame,GenomicRanges.Rdash.method}\n",
       "\\aliasA{as.factor,GenomicRanges-method}{GRanges-class}{as.factor,GenomicRanges.Rdash.method}\n",
       "\\aliasA{bindROWS,GenomicRanges-method}{GRanges-class}{bindROWS,GenomicRanges.Rdash.method}\n",
       "\\aliasA{class:GenomicPos}{GRanges-class}{class:GenomicPos}\n",
       "\\aliasA{class:GenomicRanges}{GRanges-class}{class:GenomicRanges}\n",
       "\\aliasA{class:GRanges}{GRanges-class}{class:GRanges}\n",
       "\\aliasA{class:IRanges\\_OR\\_IPos}{GRanges-class}{class:IRanges.Rul.OR.Rul.IPos}\n",
       "\\aliasA{coerce,ANY,GenomicRanges-method}{GRanges-class}{coerce,ANY,GenomicRanges.Rdash.method}\n",
       "\\aliasA{coerce,character,GRanges-method}{GRanges-class}{coerce,character,GRanges.Rdash.method}\n",
       "\\aliasA{coerce,factor,GRanges-method}{GRanges-class}{coerce,factor,GRanges.Rdash.method}\n",
       "\\aliasA{coerce,GenomicRanges,CompressedIRangesList-method}{GRanges-class}{coerce,GenomicRanges,CompressedIRangesList.Rdash.method}\n",
       "\\aliasA{coerce,GenomicRanges,GRanges-method}{GRanges-class}{coerce,GenomicRanges,GRanges.Rdash.method}\n",
       "\\aliasA{coerce,GenomicRanges,Grouping-method}{GRanges-class}{coerce,GenomicRanges,Grouping.Rdash.method}\n",
       "\\aliasA{coerce,GenomicRanges,IntegerRangesList-method}{GRanges-class}{coerce,GenomicRanges,IntegerRangesList.Rdash.method}\n",
       "\\aliasA{coerce,GenomicRanges,IRangesList-method}{GRanges-class}{coerce,GenomicRanges,IRangesList.Rdash.method}\n",
       "\\aliasA{coerce,IntegerRangesList,GRanges-method}{GRanges-class}{coerce,IntegerRangesList,GRanges.Rdash.method}\n",
       "\\aliasA{coerce,Seqinfo,GRanges-method}{GRanges-class}{coerce,Seqinfo,GRanges.Rdash.method}\n",
       "\\aliasA{coerce,Seqinfo,IntegerRangesList-method}{GRanges-class}{coerce,Seqinfo,IntegerRangesList.Rdash.method}\n",
       "\\aliasA{end<\\Rdash{},GenomicRanges-method}{GRanges-class}{end<.Rdash.,GenomicRanges.Rdash.method}\n",
       "\\aliasA{GenomicPos}{GRanges-class}{GenomicPos}\n",
       "\\aliasA{GenomicPos-class}{GRanges-class}{GenomicPos.Rdash.class}\n",
       "\\aliasA{GenomicRanges}{GRanges-class}{GenomicRanges}\n",
       "\\aliasA{GenomicRanges\\_OR\\_missing-class}{GRanges-class}{GenomicRanges.Rul.OR.Rul.missing.Rdash.class}\n",
       "\\aliasA{GenomicRanges-class}{GRanges-class}{GenomicRanges.Rdash.class}\n",
       "\\aliasA{getListElement,GenomicRanges-method}{GRanges-class}{getListElement,GenomicRanges.Rdash.method}\n",
       "\\aliasA{GRanges}{GRanges-class}{GRanges}\n",
       "\\aliasA{GRanges}{GRanges-class}{GRanges}\n",
       "\\aliasA{granges,GenomicRanges-method}{GRanges-class}{granges,GenomicRanges.Rdash.method}\n",
       "\\aliasA{IRanges\\_OR\\_IPos}{GRanges-class}{IRanges.Rul.OR.Rul.IPos}\n",
       "\\aliasA{IRanges\\_OR\\_IPos-class}{GRanges-class}{IRanges.Rul.OR.Rul.IPos.Rdash.class}\n",
       "\\aliasA{length,GenomicRanges-method}{GRanges-class}{length,GenomicRanges.Rdash.method}\n",
       "\\aliasA{names,GenomicRanges-method}{GRanges-class}{names,GenomicRanges.Rdash.method}\n",
       "\\aliasA{names<\\Rdash{},GenomicRanges-method}{GRanges-class}{names<.Rdash.,GenomicRanges.Rdash.method}\n",
       "\\aliasA{parallel\\_slot\\_names,GRanges-method}{GRanges-class}{parallel.Rul.slot.Rul.names,GRanges.Rdash.method}\n",
       "\\aliasA{ranges,GRanges-method}{GRanges-class}{ranges,GRanges.Rdash.method}\n",
       "\\aliasA{ranges<\\Rdash{},GenomicRanges-method}{GRanges-class}{ranges<.Rdash.,GenomicRanges.Rdash.method}\n",
       "\\aliasA{score,GenomicRanges-method}{GRanges-class}{score,GenomicRanges.Rdash.method}\n",
       "\\aliasA{score<\\Rdash{},GenomicRanges-method}{GRanges-class}{score<.Rdash.,GenomicRanges.Rdash.method}\n",
       "\\aliasA{seqinfo,GRanges-method}{GRanges-class}{seqinfo,GRanges.Rdash.method}\n",
       "\\aliasA{seqinfo,List-method}{GRanges-class}{seqinfo,List.Rdash.method}\n",
       "\\aliasA{seqinfo<\\Rdash{},GenomicRanges-method}{GRanges-class}{seqinfo<.Rdash.,GenomicRanges.Rdash.method}\n",
       "\\aliasA{seqinfo<\\Rdash{},List-method}{GRanges-class}{seqinfo<.Rdash.,List.Rdash.method}\n",
       "\\aliasA{seqnames,GRanges-method}{GRanges-class}{seqnames,GRanges.Rdash.method}\n",
       "\\aliasA{seqnames<\\Rdash{},GenomicRanges-method}{GRanges-class}{seqnames<.Rdash.,GenomicRanges.Rdash.method}\n",
       "\\aliasA{show,GenomicRanges-method}{GRanges-class}{show,GenomicRanges.Rdash.method}\n",
       "\\aliasA{start,GenomicRanges-method}{GRanges-class}{start,GenomicRanges.Rdash.method}\n",
       "\\aliasA{start<\\Rdash{},GenomicRanges-method}{GRanges-class}{start<.Rdash.,GenomicRanges.Rdash.method}\n",
       "\\aliasA{strand,GRanges-method}{GRanges-class}{strand,GRanges.Rdash.method}\n",
       "\\aliasA{strand<\\Rdash{},GenomicRanges,ANY-method}{GRanges-class}{strand<.Rdash.,GenomicRanges,ANY.Rdash.method}\n",
       "\\aliasA{summary,GenomicRanges-method}{GRanges-class}{summary,GenomicRanges.Rdash.method}\n",
       "\\aliasA{summary.GenomicRanges}{GRanges-class}{summary.GenomicRanges}\n",
       "\\aliasA{update,GRanges-method}{GRanges-class}{update,GRanges.Rdash.method}\n",
       "\\aliasA{updateObject,GRanges-method}{GRanges-class}{updateObject,GRanges.Rdash.method}\n",
       "\\aliasA{width,GenomicRanges-method}{GRanges-class}{width,GenomicRanges.Rdash.method}\n",
       "\\aliasA{width<\\Rdash{},GenomicRanges-method}{GRanges-class}{width<.Rdash.,GenomicRanges.Rdash.method}\n",
       "\\keyword{methods}{GRanges-class}\n",
       "\\keyword{classes}{GRanges-class}\n",
       "%\n",
       "\\begin{Description}\\relax\n",
       "The GRanges class is a container for the genomic locations and\n",
       "their associated annotations.\n",
       "\\end{Description}\n",
       "%\n",
       "\\begin{Details}\\relax\n",
       "GRanges is a vector of genomic locations and associated\n",
       "annotations. Each element in the vector is comprised of a sequence name,\n",
       "an interval, a \\LinkA{strand}{strand}, and optional metadata columns (e.g. score, GC\n",
       "content, etc.). This information is stored in four components:\n",
       "\\begin{description}\n",
       "\n",
       "\\item[\\code{seqnames}] a 'factor' \\LinkA{Rle}{Rle} object\n",
       "containing the sequence names.\n",
       "\\item[\\code{ranges}] an \\LinkA{IRanges}{IRanges} object containing\n",
       "the ranges.\n",
       "\\item[\\code{strand}] a 'factor' \\LinkA{Rle}{Rle} object containing\n",
       "the \\LinkA{strand}{strand} information.\n",
       "\\item[\\code{mcols}] a \\LinkA{DataFrame}{DataFrame} object\n",
       "containing the metadata columns. Columns cannot be named\n",
       "\\code{\"seqnames\"}, \\code{\"ranges\"}, \\code{\"strand\"},\n",
       "\\code{\"seqlevels\"}, \\code{\"seqlengths\"}, \\code{\"isCircular\"},\n",
       "\\code{\"start\"}, \\code{\"end\"}, \\code{\"width\"}, or \\code{\"element\"}.\n",
       "\\item[\\code{seqinfo}] a \\LinkA{Seqinfo}{Seqinfo} object containing information\n",
       "about the set of genomic sequences present in the GRanges object.\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "\\end{Details}\n",
       "%\n",
       "\\begin{Section}{Constructor}\n",
       "\\begin{description}\n",
       "\n",
       "\\item[] \n",
       "\\code{GRanges(seqnames=NULL, ranges=NULL, strand=NULL,\n",
       "                    ..., seqinfo=NULL, seqlengths=NULL)}:\n",
       "Creates a GRanges object.\n",
       "\\begin{description}\n",
       "\n",
       "\\item[\\code{seqnames}] \n",
       "\\code{NULL}, or an \\LinkA{Rle}{Rle} object, character vector,\n",
       "or factor containing the sequence names.\n",
       "\n",
       "\\item[\\code{ranges}] \n",
       "\\code{NULL}, or an \\LinkA{IRanges}{IRanges} object containing the\n",
       "ranges.\n",
       "\n",
       "\\item[\\code{strand}] \n",
       "\\code{NULL}, or an \\LinkA{Rle}{Rle} object, character vector,\n",
       "or factor containing the strand information.\n",
       "\n",
       "\\item[\\code{...}] \n",
       "Metadata columns to set on the GRanges object. All the metadata\n",
       "columns must be vector-like objects of the same length as the object\n",
       "to construct. They cannot be named \\code{\"start\"}, \\code{\"end\"},\n",
       "\\code{\"width\"}, or \\code{\"element\"}.\n",
       "\n",
       "\\item[\\code{seqinfo}] \n",
       "Either \\code{NULL}, or a \\LinkA{Seqinfo}{Seqinfo} object,\n",
       "or a character vector of unique sequence names (a.k.a.\n",
       "\\emph{seqlevels}), or a named numeric vector of sequence lengths.\n",
       "When not \\code{NULL}, \\code{seqinfo} must be compatible with the\n",
       "sequence names in \\code{seqnames}, that is, it must have one entry\n",
       "for each unique sequence name in \\code{seqnames}. Note that it can\n",
       "have additional entries i.e. entries for seqlevels not present\n",
       "in \\code{seqnames}.\n",
       "\n",
       "\\item[\\code{seqlengths}] \n",
       "\\code{NULL}, or an integer vector named with \\code{levels(seqnames)}\n",
       "and containing the lengths (or NA) for each level in\n",
       "\\code{levels(seqnames)}.\n",
       "\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "If \\code{ranges} is not supplied and/or NULL then the constructor\n",
       "proceeds in 2 steps:\n",
       "\\begin{enumerate}\n",
       "\n",
       "\\item{} An initial GRanges object is created with\n",
       "\\code{as(seqnames, \"GRanges\")}.\n",
       "\\item{} Then this GRanges object is updated according to whatever\n",
       "non-NULL remaining arguments were passed to the call to\n",
       "\\code{GRanges()}.\n",
       "\n",
       "\\end{enumerate}\n",
       "\n",
       "As a consequence of this behavior, \\code{GRanges(x)} is equivalent to\n",
       "\\code{as(x, \"GRanges\")}.\n",
       "\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "\\end{Section}\n",
       "%\n",
       "\\begin{Section}{Accessors}\n",
       "In the following code snippets, \\code{x} is a GRanges object.\n",
       "\n",
       "\\begin{description}\n",
       "\n",
       "\\item[] \n",
       "\\code{length(x)}:\n",
       "Get the number of elements.\n",
       "\n",
       "\\item[] \n",
       "\\code{seqnames(x)}, \\code{seqnames(x) <- value}:\n",
       "Get or set the sequence names.\n",
       "\\code{value} can be an \\LinkA{Rle}{Rle} object, a character vector,\n",
       "or a factor.\n",
       "\n",
       "\\item[] \n",
       "\\code{ranges(x)}, \\code{ranges(x) <- value}:\n",
       "Get or set the ranges. \\code{value} can be an\n",
       "\\LinkA{IntegerRanges}{IntegerRanges} object.\n",
       "\n",
       "\\item[] \n",
       "\\code{start(x)}, \\code{start(x) <- value}:\n",
       "Get or set \\code{start(ranges(x))}.\n",
       "\n",
       "\\item[] \n",
       "\\code{end(x)}, \\code{end(x) <- value}:\n",
       "Get or set \\code{end(ranges(x))}.\n",
       "\n",
       "\\item[] \n",
       "\\code{width(x)}, \\code{width(x) <- value}:\n",
       "Get or set \\code{width(ranges(x))}.\n",
       "\n",
       "\\item[] \n",
       "\\code{strand(x)}, \\code{strand(x) <- value}:\n",
       "Get or set the strand. \\code{value} can be an Rle object, character\n",
       "vector, or factor.\n",
       "\n",
       "\\item[] \n",
       "\\code{names(x)}, \\code{names(x) <- value}:\n",
       "Get or set the names of the elements.\n",
       "\n",
       "\\item[] \n",
       "\\code{mcols(x, use.names=FALSE)}, \\code{mcols(x) <- value}:\n",
       "Get or set the metadata columns.\n",
       "If \\code{use.names=TRUE} and the metadata columns are not \\code{NULL},\n",
       "then the names of \\code{x} are propagated as the row names of the\n",
       "returned \\LinkA{DataFrame}{DataFrame} object.\n",
       "When setting the metadata columns, the supplied value must be \\code{NULL}\n",
       "or a data-frame-like object (i.e. \\LinkA{DataFrame}{DataFrame} or data.frame)\n",
       "holding element-wise metadata.\n",
       "\n",
       "\\item[] \n",
       "\\code{elementMetadata(x)}, \\code{elementMetadata(x) <- value},\n",
       "\\code{values(x)}, \\code{values(x) <- value}:\n",
       "Alternatives to \\code{mcols} functions. Their use is discouraged.\n",
       "\n",
       "\\item[] \n",
       "\\code{seqinfo(x)}, \\code{seqinfo(x) <- value}:\n",
       "Get or set the information about the underlying sequences.\n",
       "\\code{value} must be a \\LinkA{Seqinfo}{Seqinfo} object.\n",
       "\n",
       "\\item[] \n",
       "\\code{seqlevels(x)},\n",
       "\\code{seqlevels(x, pruning.mode=c(\"error\", \"coarse\", \"fine\", \"tidy\")) <- value}:\n",
       "Get or set the sequence levels.\n",
       "\\code{seqlevels(x)} is equivalent to \\code{seqlevels(seqinfo(x))}\n",
       "or to \\code{levels(seqnames(x))}, those 2 expressions being\n",
       "guaranteed to return identical character vectors on a GRanges object.\n",
       "\\code{value} must be a character vector with no NAs.\n",
       "See \\code{?\\LinkA{seqlevels}{seqlevels}} for more information.\n",
       "\n",
       "\\item[] \n",
       "\\code{seqlengths(x)}, \\code{seqlengths(x) <- value}:\n",
       "Get or set the sequence lengths.\n",
       "\\code{seqlengths(x)} is equivalent to \\code{seqlengths(seqinfo(x))}.\n",
       "\\code{value} can be a named non-negative integer or numeric vector\n",
       "eventually with NAs.\n",
       "\n",
       "\\item[] \n",
       "\\code{isCircular(x)}, \\code{isCircular(x) <- value}:\n",
       "Get or set the circularity flags.\n",
       "\\code{isCircular(x)} is equivalent to \\code{isCircular(seqinfo(x))}.\n",
       "\\code{value} must be a named logical vector eventually with NAs.\n",
       "\n",
       "\\item[] \n",
       "\\code{genome(x)}, \\code{genome(x) <- value}:\n",
       "Get or set the genome identifier or assembly name for each sequence.\n",
       "\\code{genome(x)} is equivalent to \\code{genome(seqinfo(x))}.\n",
       "\\code{value} must be a named character vector eventually with NAs.\n",
       "\n",
       "\\item[] \n",
       "\\code{seqlevelsStyle(x)}, \\code{seqlevelsStyle(x) <- value}:\n",
       "Get or set the seqname style for \\code{x}.\n",
       "See the \\LinkA{seqlevelsStyle}{seqlevelsStyle} generic getter and setter\n",
       "in the \\pkg{GenomeInfoDb} package for more information.\n",
       "\n",
       "\\item[] \n",
       "\\code{score(x), score(x) <- value}: Get or set the ``score''\n",
       "column from the element metadata.\n",
       "\n",
       "\\item[] \n",
       "\\code{granges(x, use.names=FALSE, use.mcols=FALSE)}: Squeeze the genomic\n",
       "ranges out of \\LinkA{GenomicRanges}{GenomicRanges} object \\code{x} and return them in a\n",
       "GRanges object \\emph{parallel} to \\code{x} (i.e. same length as \\code{x}).\n",
       "If \\code{use.mcols} is \\code{TRUE}, the metadata columns are propagated.\n",
       "If \\code{x} is a \\LinkA{GenomicRanges}{GenomicRanges} derivative with \\emph{extra column\n",
       "slots}, these will be propagated as metadata columns on the returned\n",
       "GRanges object.\n",
       "\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "\\end{Section}\n",
       "%\n",
       "\\begin{Section}{Coercion}\n",
       "In the code snippets below, \\code{x} is a GRanges object.\n",
       "\n",
       "\\begin{description}\n",
       "\n",
       "\\item[] \n",
       "\\code{as(from, \"GRanges\")}: Creates a GRanges object from a character\n",
       "vector, a factor, or \\LinkA{IntegerRangesList}{IntegerRangesList} object.\n",
       "\n",
       "When \\code{from} is a character vector (or a factor), each element\n",
       "in it must represent a genomic range in format \\code{chr1:2501-2800}\n",
       "(unstranded range) or \\code{chr1:2501-2800:+} (stranded range).\n",
       "\\code{..} is also supported as a separator between the start and end\n",
       "positions. Strand can be \\code{+}, \\code{-}, \\code{*}, or missing.\n",
       "The names on \\code{from} are propagated to the returned GRanges object.\n",
       "See \\code{as.character()} and \\code{as.factor()} below for the\n",
       "reverse transformations.\n",
       "\n",
       "Coercing a data.frame or DataFrame into a GRanges object is also\n",
       "supported. See \\code{\\LinkA{makeGRangesFromDataFrame}{makeGRangesFromDataFrame}} for the details.\n",
       "\n",
       "\\item[] \n",
       "\\code{as(from, \"IntegerRangesList\")}:\n",
       "Creates a \\LinkA{IntegerRangesList}{IntegerRangesList} object from a GRanges\n",
       "object. The \\code{strand} and metadata columns become \\emph{inner}\n",
       "metadata columns (i.e. metadata columns on the ranges).\n",
       "The \\code{seqlengths(from)}, \\code{isCircular(from)}, and\n",
       "\\code{genome(from)} vectors become the metadata columns.\n",
       "\n",
       "\\item[] \n",
       "\\code{as.character(x, ignore.strand=FALSE)}:\n",
       "Turn GRanges object \\code{x} into a character vector where each\n",
       "range in \\code{x} is represented by a string in format\n",
       "\\code{chr1:2501-2800:+}. If \\code{ignore.strand} is TRUE or if\n",
       "\\emph{all} the ranges in \\code{x} are unstranded (i.e. their strand\n",
       "is set to \\code{*}), then all the strings in the output are in\n",
       "format \\code{chr1:2501-2800}.\n",
       "\n",
       "The names on \\code{x} are propagated to the returned character vector.\n",
       "Its metadata (\\code{metadata(x)}) and metadata columns (\\code{mcols(x)})\n",
       "are ignored.\n",
       "\n",
       "See \\code{as(from, \"GRanges\")} above for the reverse transformation.\n",
       "\n",
       "\\item[] \n",
       "\\code{as.factor(x)}:\n",
       "Equivalent to\n",
       "\\begin{alltt}  factor(as.character(x), levels=as.character(sort(unique(x))))\n",
       "\\end{alltt}\n",
       "\n",
       "See \\code{as(from, \"GRanges\")} above for the reverse transformation.\n",
       "\n",
       "Note that \\code{table(x)} is supported on a GRanges object. It is\n",
       "equivalent to, but much faster than, \\code{table(as.factor(x))}.\n",
       "\n",
       "\\item[] \n",
       "\\code{as.data.frame(x, row.names = NULL, optional = FALSE, ...)}:\n",
       "Creates a data.frame with columns \\code{seqnames} (factor),\n",
       "\\code{start} (integer), \\code{end} (integer), \\code{width} (integer),\n",
       "\\code{strand} (factor), as well as the additional metadata columns\n",
       "stored in \\code{mcols(x)}. Pass an explicit\n",
       "\\code{stringsAsFactors=TRUE/FALSE} argument via \\code{...} to\n",
       "override the default conversions for the metadata columns in\n",
       "\\code{mcols(x)}.\n",
       "\n",
       "\\item[] \n",
       "\\code{as(from, \"Grouping\")}: Creates a\n",
       "\\code{\\LinkA{ManyToOneGrouping}{ManyToOneGrouping}} object that groups\n",
       "\\code{from} by seqname, strand, start and end (same as the default\n",
       "sort order). This makes it convenient, for example, to aggregate a\n",
       "GenomicRanges object by range.\n",
       "\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "\n",
       "In the code snippets below, \\code{x} is a \\LinkA{Seqinfo}{Seqinfo}\n",
       "object.\n",
       "\n",
       "\\begin{description}\n",
       "\n",
       "\\item[] \n",
       "\\code{as(x, \"GRanges\")}, \\code{as(x, \"GenomicRanges\")},\n",
       "\\code{as(x, \"IntegerRangesList\")}: Turns \\LinkA{Seqinfo}{Seqinfo}\n",
       "object \\code{x} (with no \\code{NA} lengths) into a GRanges or\n",
       "\\LinkA{IntegerRangesList}{IntegerRangesList}.\n",
       "\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "\\end{Section}\n",
       "%\n",
       "\\begin{Section}{Subsetting}\n",
       "In the code snippets below, \\code{x} is a GRanges object.\n",
       "\\begin{description}\n",
       "\n",
       "\\item[] \n",
       "\\code{x[i]}:\n",
       "Return a new GRanges object made of the elements selected by \\code{i}.\n",
       "\n",
       "\\item[] \n",
       "\\code{x[i, j]}:\n",
       "Like the above, but allow the user to conveniently subset the metadata\n",
       "columns thru \\code{j}.\n",
       "\n",
       "\\item[] \n",
       "\\code{x[i] <- value}:\n",
       "Replacement version of \\code{x[i]}.\n",
       "\n",
       "\\item[] \n",
       "\\code{x\\$name}, \\code{x\\$name <- value}:\n",
       "Shortcuts for \\code{mcols(x)\\$name} and \\code{mcols(x)\\$name <- value},\n",
       "respectively. Provided as a convenience, for GRanges objects *only*,\n",
       "and as the result of strong popular demand.\n",
       "Note that those methods are not consistent with the other \\code{\\$}\n",
       "and \\code{\\$<-} methods in the IRanges/GenomicRanges infrastructure,\n",
       "and might confuse some users by making them believe that a GRanges\n",
       "object can be manipulated as a data.frame-like object.\n",
       "Therefore we recommend using them only interactively, and we discourage\n",
       "their use in scripts or packages. For the latter, use\n",
       "\\code{mcols(x)\\$name} and \\code{mcols(x)\\$name <- value}, instead\n",
       "of \\code{x\\$name} and \\code{x\\$name <- value}, respectively.\n",
       "\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "\n",
       "See \\code{?`\\LinkA{[}{[}`} in the \\pkg{S4Vectors} package for more\n",
       "information about subsetting Vector derivatives and for an important note\n",
       "about the \\code{x[i, j]} form.\n",
       "\n",
       "Note that a GRanges object can be used as a subscript to subset a\n",
       "list-like object that has names on it. In that case, the names on the\n",
       "list-like object are interpreted as sequence names.\n",
       "In the code snippets below, \\code{x} is a list or \\LinkA{List}{List} object with\n",
       "names on it, and the subscript \\code{gr} is a GRanges object with all its\n",
       "seqnames being valid \\code{x} names.\n",
       "\n",
       "\\begin{description}\n",
       "\n",
       "\\item[] \n",
       "\\code{x[gr]}:\n",
       "Return an object of the same class as \\code{x} and \\emph{parallel}\n",
       "to \\code{gr}. More precisely, it's conceptually doing:\n",
       "\\begin{alltt}  lapply(gr, function(gr1) x[[seqnames(gr1)]][ranges(gr1)])\n",
       "\\end{alltt}\n",
       "\n",
       "\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "\\end{Section}\n",
       "%\n",
       "\\begin{Section}{Concatenation}\n",
       "\\begin{description}\n",
       "\n",
       "\\item[] \n",
       "\\code{c(x, ..., ignore.mcols=FALSE)}:\n",
       "Concatenate GRanges object \\code{x} and the GRanges objects in\n",
       "\\code{...} together.\n",
       "See \\code{?\\LinkA{c}{c}} in the \\pkg{S4Vectors} package\n",
       "for more information about concatenating Vector derivatives.\n",
       "\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "\\end{Section}\n",
       "%\n",
       "\\begin{Section}{Splitting}\n",
       "\\begin{description}\n",
       "\n",
       "\\item[] \n",
       "\\code{split(x, f, drop=FALSE)}:\n",
       "Splits GRanges object \\code{x} according to \\code{f} to create a\n",
       "\\LinkA{GRangesList}{GRangesList} object.\n",
       "If \\code{f} is a list-like object then \\code{drop} is ignored\n",
       "and \\code{f} is treated as if it was\n",
       "\\code{rep(seq\\_len(length(f)), sapply(f, length))},\n",
       "so the returned object has the same shape as \\code{f} (it also\n",
       "receives the names of \\code{f}).\n",
       "Otherwise, if \\code{f} is not a list-like object, empty list\n",
       "elements are removed from the returned object if \\code{drop} is\n",
       "\\code{TRUE}.\n",
       "\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "\\end{Section}\n",
       "%\n",
       "\\begin{Section}{Displaying}\n",
       "In the code snippets below, \\code{x} is a GRanges object.\n",
       "\n",
       "\\begin{description}\n",
       "\n",
       "\\item[] \n",
       "\\code{show(x)}:\n",
       "By default the \\code{show} method displays 5 head and 5 tail\n",
       "elements. This can be changed by setting the global options\n",
       "\\code{showHeadLines} and \\code{showTailLines}. If the object\n",
       "length is less than (or equal to) the sum of these 2 options\n",
       "plus 1, then the full object is displayed.\n",
       "Note that these options also affect the display of\n",
       "\\LinkA{GAlignments}{GAlignments} and\n",
       "\\LinkA{GAlignmentPairs}{GAlignmentPairs} objects (defined in\n",
       "the \\pkg{GenomicAlignments} package), as well as other objects\n",
       "defined in the \\pkg{IRanges} and \\pkg{Biostrings} packages (e.g.\n",
       "\\LinkA{IRanges}{IRanges} and \\LinkA{DNAStringSet}{DNAStringSet} objects).\n",
       "\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "\\end{Section}\n",
       "%\n",
       "\\begin{Author}\\relax\n",
       "P. Aboyoun and H. Pag√®s\n",
       "\\end{Author}\n",
       "%\n",
       "\\begin{SeeAlso}\\relax\n",
       "\\begin{itemize}\n",
       "\n",
       "\\item{} The \\LinkA{IRanges}{IRanges} class in the \\pkg{IRanges}\n",
       "package for storing a set of \\emph{integer ranges}.\n",
       "\n",
       "\\item{} The \\LinkA{GPos}{GPos} class for representing a set of \\emph{genomic\n",
       "positions} (i.e. \\emph{genomic ranges} of width 1, a.k.a.\n",
       "\\emph{genomic loci}).\n",
       "\n",
       "\\item{} \\code{\\LinkA{makeGRangesFromDataFrame}{makeGRangesFromDataFrame}} for making a GRanges object\n",
       "from a data.frame or \\LinkA{DataFrame}{DataFrame} object.\n",
       "\n",
       "\\item{} \\LinkA{Seqinfo}{Seqinfo} objects and the\n",
       "\\code{\\LinkA{seqinfo}{seqinfo}} accessor and family in the\n",
       "\\pkg{GenomeInfoDb} package for accessing/modifying information\n",
       "about the underlying sequences of a \\LinkA{GenomicRanges}{GenomicRanges} derivative.\n",
       "\n",
       "\\item{} \\LinkA{GenomicRanges-comparison}{GenomicRanges.Rdash.comparison} for comparing and ordering genomic\n",
       "ranges and/or positions.\n",
       "\n",
       "\\item{} \\LinkA{findOverlaps-methods}{findOverlaps.Rdash.methods} for finding overlapping\n",
       "genomic ranges and/or positions.\n",
       "\n",
       "\\item{} \\LinkA{intra-range-methods}{intra.Rdash.range.Rdash.methods} and\n",
       "\\LinkA{inter-range-methods}{inter.Rdash.range.Rdash.methods}\n",
       "for \\emph{intra range} and \\emph{inter range} transformations\n",
       "of \\LinkA{GenomicRanges}{GenomicRanges} derivatives.\n",
       "\n",
       "\\item{} \\LinkA{coverage-methods}{coverage.Rdash.methods} for computing the coverage\n",
       "of a set of genomic ranges and/or positions.\n",
       "\n",
       "\\item{} \\LinkA{setops-methods}{setops.Rdash.methods} for set operations on GRanges\n",
       "objects.\n",
       "\n",
       "\\item{} \\LinkA{nearest-methods}{nearest.Rdash.methods} for finding the nearest\n",
       "genomic range/position neighbor.\n",
       "\n",
       "\\item{} \\code{\\LinkA{absoluteRanges}{absoluteRanges}} for transforming genomic ranges into\n",
       "\\emph{absolute} ranges (i.e. into ranges on the sequence obtained\n",
       "by virtually concatenating all the sequences in a genome).\n",
       "\n",
       "\\item{} \\code{\\LinkA{tileGenome}{tileGenome}} for putting tiles on a genome.\n",
       "\n",
       "\\item{} \\LinkA{genomicvars}{genomicvars} for manipulating genomic variables.\n",
       "\n",
       "\\item{} \\LinkA{GRangesList}{GRangesList} objects.\n",
       "\n",
       "\\item{} \\LinkA{Vector}{Vector}, \\LinkA{Rle}{Rle}, and\n",
       "\\LinkA{DataFrame}{DataFrame} objects in the \\pkg{S4Vectors} package.\n",
       "\n",
       "\\end{itemize}\n",
       "\n",
       "\\end{SeeAlso}\n",
       "%\n",
       "\\begin{Examples}\n",
       "\\begin{ExampleCode}\n",
       "showClass(\"GRanges\")  # shows the known subclasses\n",
       "\n",
       "## ---------------------------------------------------------------------\n",
       "## CONSTRUCTION\n",
       "## ---------------------------------------------------------------------\n",
       "## Specifying the bare minimum i.e. seqnames and ranges only. The\n",
       "## GRanges object will have no names, no strand information, and no\n",
       "## metadata columns:\n",
       "gr0 <- GRanges(Rle(c(\"chr2\", \"chr2\", \"chr1\", \"chr3\"), c(1, 3, 2, 4)),\n",
       "               IRanges(1:10, width=10:1))\n",
       "gr0\n",
       "\n",
       "## Specifying names, strand, metadata columns. They can be set on an\n",
       "## existing object:\n",
       "names(gr0) <- head(letters, 10)\n",
       "strand(gr0) <- Rle(strand(c(\"-\", \"+\", \"*\", \"+\", \"-\")), c(1, 2, 2, 3, 2))\n",
       "mcols(gr0)$score <- 1:10\n",
       "mcols(gr0)$GC <- seq(1, 0, length=10)\n",
       "gr0\n",
       "\n",
       "## ... or specified at construction time:\n",
       "gr <- GRanges(Rle(c(\"chr2\", \"chr2\", \"chr1\", \"chr3\"), c(1, 3, 2, 4)),\n",
       "              IRanges(1:10, width=10:1, names=head(letters, 10)),\n",
       "              Rle(strand(c(\"-\", \"+\", \"*\", \"+\", \"-\")), c(1, 2, 2, 3, 2)),\n",
       "              score=1:10, GC=seq(1, 0, length=10))\n",
       "stopifnot(identical(gr0, gr))\n",
       "\n",
       "## Specifying the seqinfo. It can be set on an existing object:\n",
       "seqinfo <- Seqinfo(paste0(\"chr\", 1:3), c(1000, 2000, 1500), NA, \"mock1\")\n",
       "seqinfo(gr0) <- merge(seqinfo(gr0), seqinfo)\n",
       "seqlevels(gr0) <- seqlevels(seqinfo)\n",
       "\n",
       "## ... or specified at construction time:\n",
       "gr <- GRanges(Rle(c(\"chr2\", \"chr2\", \"chr1\", \"chr3\"), c(1, 3, 2, 4)),\n",
       "              IRanges(1:10, width=10:1, names=head(letters, 10)),\n",
       "              Rle(strand(c(\"-\", \"+\", \"*\", \"+\", \"-\")), c(1, 2, 2, 3, 2)),\n",
       "              score=1:10, GC=seq(1, 0, length=10),\n",
       "              seqinfo=seqinfo)\n",
       "stopifnot(identical(gr0, gr))\n",
       "\n",
       "## ---------------------------------------------------------------------\n",
       "## COERCION\n",
       "## ---------------------------------------------------------------------\n",
       "## From GRanges:\n",
       "as.character(gr)\n",
       "as.factor(gr)\n",
       "as.data.frame(gr)\n",
       "\n",
       "## From character to GRanges:\n",
       "x1 <- \"chr2:56-125\"\n",
       "as(x1, \"GRanges\")\n",
       "as(rep(x1, 4), \"GRanges\")\n",
       "x2 <- c(A=x1, B=\"chr1:25-30:-\")\n",
       "as(x2, \"GRanges\")\n",
       "\n",
       "## From data.frame to GRanges:\n",
       "df <- data.frame(chrom=\"chr2\", start=11:15, end=20:24)\n",
       "gr3 <- as(df, \"GRanges\")\n",
       "\n",
       "## Alternatively, coercion to GRanges can be done by just calling the\n",
       "## GRanges() constructor on the object to coerce:\n",
       "gr1 <- GRanges(x1)  # same as as(x1, \"GRanges\")\n",
       "gr2 <- GRanges(x2)  # same as as(x2, \"GRanges\")\n",
       "gr3 <- GRanges(df)  # same as as(df, \"GRanges\")\n",
       "\n",
       "## Sanity checks:\n",
       "stopifnot(identical(as(x1, \"GRanges\"), gr1))\n",
       "stopifnot(identical(as(x2, \"GRanges\"), gr2))\n",
       "stopifnot(identical(as(df, \"GRanges\"), gr3))\n",
       "\n",
       "## ---------------------------------------------------------------------\n",
       "## SUMMARIZING ELEMENTS\n",
       "## ---------------------------------------------------------------------\n",
       "table(seqnames(gr))\n",
       "table(strand(gr))\n",
       "sum(width(gr))\n",
       "table(gr)\n",
       "summary(mcols(gr)[,\"score\"])\n",
       "\n",
       "## The number of lines displayed in the 'show' method are controlled\n",
       "## with two global options:\n",
       "longGR  <- sample(gr, 25, replace=TRUE)\n",
       "longGR\n",
       "options(showHeadLines=7)\n",
       "options(showTailLines=2)\n",
       "longGR\n",
       "\n",
       "## Revert to default values\n",
       "options(showHeadLines=NULL)\n",
       "options(showTailLines=NULL)\n",
       "\n",
       "## ---------------------------------------------------------------------\n",
       "## INVERTING THE STRAND\n",
       "## ---------------------------------------------------------------------\n",
       "invertStrand(gr)\n",
       "\n",
       "## ---------------------------------------------------------------------\n",
       "## RENAMING THE UNDERLYING SEQUENCES\n",
       "## ---------------------------------------------------------------------\n",
       "seqlevels(gr)\n",
       "seqlevels(gr) <- sub(\"chr\", \"Chrom\", seqlevels(gr))\n",
       "gr\n",
       "seqlevels(gr) <- sub(\"Chrom\", \"chr\", seqlevels(gr)) # revert\n",
       "\n",
       "## ---------------------------------------------------------------------\n",
       "## COMBINING OBJECTS\n",
       "## ---------------------------------------------------------------------\n",
       "gr2 <- GRanges(seqnames=Rle(c('chr1', 'chr2', 'chr3'), c(3, 3, 4)),\n",
       "               IRanges(1:10, width=5),\n",
       "               strand='-',\n",
       "               score=101:110, GC=runif(10),\n",
       "               seqinfo=seqinfo)\n",
       "gr3 <- GRanges(seqnames=Rle(c('chr1', 'chr2', 'chr3'), c(3, 4, 3)),\n",
       "               IRanges(101:110, width=10),\n",
       "               strand='-',\n",
       "               score=21:30,\n",
       "               seqinfo=seqinfo)\n",
       "some.gr <- c(gr, gr2)\n",
       "\n",
       "c(gr, gr2, gr3)\n",
       "c(gr, gr2, gr3, ignore.mcols=TRUE)\n",
       "\n",
       "## ---------------------------------------------------------------------\n",
       "## USING A GRANGES OBJECT AS A SUBSCRIPT TO SUBSET ANOTHER OBJECT\n",
       "## ---------------------------------------------------------------------\n",
       "## Subsetting *by* a GRanges subscript is supported only if the object\n",
       "## to subset is a named list-like object:\n",
       "x <- RleList(chr1=101:120, chr2=2:-8, chr3=31:40)\n",
       "x[gr]\n",
       "\\end{ExampleCode}\n",
       "\\end{Examples}"
      ],
      "text/plain": [
       "GRanges-class          package:GenomicRanges           R Documentation\n",
       "\n",
       "_\bG_\bR_\ba_\bn_\bg_\be_\bs _\bo_\bb_\bj_\be_\bc_\bt_\bs\n",
       "\n",
       "_\bD_\be_\bs_\bc_\br_\bi_\bp_\bt_\bi_\bo_\bn:\n",
       "\n",
       "     The GRanges class is a container for the genomic locations and\n",
       "     their associated annotations.\n",
       "\n",
       "_\bD_\be_\bt_\ba_\bi_\bl_\bs:\n",
       "\n",
       "     GRanges is a vector of genomic locations and associated\n",
       "     annotations. Each element in the vector is comprised of a sequence\n",
       "     name, an interval, a strand, and optional metadata columns (e.g.\n",
       "     score, GC content, etc.). This information is stored in four\n",
       "     components:\n",
       "\n",
       "     ‚Äòseqnames‚Äô a 'factor' Rle object containing the sequence names.\n",
       "\n",
       "     ‚Äòranges‚Äô an IRanges object containing the ranges.\n",
       "\n",
       "     ‚Äòstrand‚Äô a 'factor' Rle object containing the strand information.\n",
       "\n",
       "     ‚Äòmcols‚Äô a DataFrame object containing the metadata columns.\n",
       "          Columns cannot be named ‚Äò\"seqnames\"‚Äô, ‚Äò\"ranges\"‚Äô, ‚Äò\"strand\"‚Äô,\n",
       "          ‚Äò\"seqlevels\"‚Äô, ‚Äò\"seqlengths\"‚Äô, ‚Äò\"isCircular\"‚Äô, ‚Äò\"start\"‚Äô,\n",
       "          ‚Äò\"end\"‚Äô, ‚Äò\"width\"‚Äô, or ‚Äò\"element\"‚Äô.\n",
       "\n",
       "     ‚Äòseqinfo‚Äô a Seqinfo object containing information about the set of\n",
       "          genomic sequences present in the GRanges object.\n",
       "\n",
       "_\bC_\bo_\bn_\bs_\bt_\br_\bu_\bc_\bt_\bo_\br:\n",
       "\n",
       "‚ÄòGRanges(seqnames=NULL, ranges=NULL, strand=NULL, ..., seqinfo=NULL,\n",
       "          seqlengths=NULL)‚Äô: Creates a GRanges object.\n",
       "\n",
       "          ‚Äòseqnames‚Äô ‚ÄòNULL‚Äô, or an Rle object, character vector, or\n",
       "              factor containing the sequence names.\n",
       "\n",
       "          ‚Äòranges‚Äô ‚ÄòNULL‚Äô, or an IRanges object containing the ranges.\n",
       "\n",
       "          ‚Äòstrand‚Äô ‚ÄòNULL‚Äô, or an Rle object, character vector, or\n",
       "              factor containing the strand information.\n",
       "\n",
       "          ‚Äò...‚Äô Metadata columns to set on the GRanges object. All the\n",
       "              metadata columns must be vector-like objects of the same\n",
       "              length as the object to construct. They cannot be named\n",
       "              ‚Äò\"start\"‚Äô, ‚Äò\"end\"‚Äô, ‚Äò\"width\"‚Äô, or ‚Äò\"element\"‚Äô.\n",
       "\n",
       "          ‚Äòseqinfo‚Äô Either ‚ÄòNULL‚Äô, or a Seqinfo object, or a character\n",
       "              vector of unique sequence names (a.k.a.  _seqlevels_), or\n",
       "              a named numeric vector of sequence lengths.  When not\n",
       "              ‚ÄòNULL‚Äô, ‚Äòseqinfo‚Äô must be compatible with the sequence\n",
       "              names in ‚Äòseqnames‚Äô, that is, it must have one entry for\n",
       "              each unique sequence name in ‚Äòseqnames‚Äô. Note that it can\n",
       "              have additional entries i.e. entries for seqlevels not\n",
       "              present in ‚Äòseqnames‚Äô.\n",
       "\n",
       "          ‚Äòseqlengths‚Äô ‚ÄòNULL‚Äô, or an integer vector named with\n",
       "              ‚Äòlevels(seqnames)‚Äô and containing the lengths (or NA) for\n",
       "              each level in ‚Äòlevels(seqnames)‚Äô.\n",
       "\n",
       "          If ‚Äòranges‚Äô is not supplied and/or NULL then the constructor\n",
       "          proceeds in 2 steps:\n",
       "\n",
       "           1. An initial GRanges object is created with ‚Äòas(seqnames,\n",
       "              \"GRanges\")‚Äô.\n",
       "\n",
       "           2. Then this GRanges object is updated according to whatever\n",
       "              non-NULL remaining arguments were passed to the call to\n",
       "              ‚ÄòGRanges()‚Äô.\n",
       "\n",
       "          As a consequence of this behavior, ‚ÄòGRanges(x)‚Äô is equivalent\n",
       "          to ‚Äòas(x, \"GRanges\")‚Äô.\n",
       "\n",
       "_\bA_\bc_\bc_\be_\bs_\bs_\bo_\br_\bs:\n",
       "\n",
       "     In the following code snippets, ‚Äòx‚Äô is a GRanges object.\n",
       "\n",
       "‚Äòlength(x)‚Äô: Get the number of elements.\n",
       "\n",
       "‚Äòseqnames(x)‚Äô, ‚Äòseqnames(x) <- value‚Äô: Get or set the sequence names.\n",
       "          ‚Äòvalue‚Äô can be an Rle object, a character vector, or a\n",
       "          factor.\n",
       "\n",
       "‚Äòranges(x)‚Äô, ‚Äòranges(x) <- value‚Äô: Get or set the ranges. ‚Äòvalue‚Äô can\n",
       "          be an IntegerRanges object.\n",
       "\n",
       "‚Äòstart(x)‚Äô, ‚Äòstart(x) <- value‚Äô: Get or set ‚Äòstart(ranges(x))‚Äô.\n",
       "\n",
       "‚Äòend(x)‚Äô, ‚Äòend(x) <- value‚Äô: Get or set ‚Äòend(ranges(x))‚Äô.\n",
       "\n",
       "‚Äòwidth(x)‚Äô, ‚Äòwidth(x) <- value‚Äô: Get or set ‚Äòwidth(ranges(x))‚Äô.\n",
       "\n",
       "‚Äòstrand(x)‚Äô, ‚Äòstrand(x) <- value‚Äô: Get or set the strand. ‚Äòvalue‚Äô can\n",
       "          be an Rle object, character vector, or factor.\n",
       "\n",
       "‚Äònames(x)‚Äô, ‚Äònames(x) <- value‚Äô: Get or set the names of the elements.\n",
       "\n",
       "‚Äòmcols(x, use.names=FALSE)‚Äô, ‚Äòmcols(x) <- value‚Äô: Get or set the\n",
       "          metadata columns.  If ‚Äòuse.names=TRUE‚Äô and the metadata\n",
       "          columns are not ‚ÄòNULL‚Äô, then the names of ‚Äòx‚Äô are propagated\n",
       "          as the row names of the returned DataFrame object.  When\n",
       "          setting the metadata columns, the supplied value must be\n",
       "          ‚ÄòNULL‚Äô or a data-frame-like object (i.e. DataFrame or\n",
       "          data.frame) holding element-wise metadata.\n",
       "\n",
       "‚ÄòelementMetadata(x)‚Äô, ‚ÄòelementMetadata(x) <- value‚Äô, ‚Äòvalues(x)‚Äô,\n",
       "          ‚Äòvalues(x) <- value‚Äô: Alternatives to ‚Äòmcols‚Äô functions.\n",
       "          Their use is discouraged.\n",
       "\n",
       "‚Äòseqinfo(x)‚Äô, ‚Äòseqinfo(x) <- value‚Äô: Get or set the information about\n",
       "          the underlying sequences.  ‚Äòvalue‚Äô must be a Seqinfo object.\n",
       "\n",
       "‚Äòseqlevels(x)‚Äô, ‚Äòseqlevels(x, pruning.mode=c(\"error\", \"coarse\", \"fine\",\n",
       "          \"tidy\")) <- value‚Äô: Get or set the sequence levels.\n",
       "          ‚Äòseqlevels(x)‚Äô is equivalent to ‚Äòseqlevels(seqinfo(x))‚Äô or to\n",
       "          ‚Äòlevels(seqnames(x))‚Äô, those 2 expressions being guaranteed\n",
       "          to return identical character vectors on a GRanges object.\n",
       "          ‚Äòvalue‚Äô must be a character vector with no NAs.  See\n",
       "          ‚Äò?seqlevels‚Äô for more information.\n",
       "\n",
       "‚Äòseqlengths(x)‚Äô, ‚Äòseqlengths(x) <- value‚Äô: Get or set the sequence\n",
       "          lengths.  ‚Äòseqlengths(x)‚Äô is equivalent to\n",
       "          ‚Äòseqlengths(seqinfo(x))‚Äô.  ‚Äòvalue‚Äô can be a named\n",
       "          non-negative integer or numeric vector eventually with NAs.\n",
       "\n",
       "‚ÄòisCircular(x)‚Äô, ‚ÄòisCircular(x) <- value‚Äô: Get or set the circularity\n",
       "          flags.  ‚ÄòisCircular(x)‚Äô is equivalent to\n",
       "          ‚ÄòisCircular(seqinfo(x))‚Äô.  ‚Äòvalue‚Äô must be a named logical\n",
       "          vector eventually with NAs.\n",
       "\n",
       "‚Äògenome(x)‚Äô, ‚Äògenome(x) <- value‚Äô: Get or set the genome identifier or\n",
       "          assembly name for each sequence.  ‚Äògenome(x)‚Äô is equivalent\n",
       "          to ‚Äògenome(seqinfo(x))‚Äô.  ‚Äòvalue‚Äô must be a named character\n",
       "          vector eventually with NAs.\n",
       "\n",
       "‚ÄòseqlevelsStyle(x)‚Äô, ‚ÄòseqlevelsStyle(x) <- value‚Äô: Get or set the\n",
       "          seqname style for ‚Äòx‚Äô.  See the seqlevelsStyle generic getter\n",
       "          and setter in the ‚ÄòGenomeInfoDb‚Äô package for more\n",
       "          information.\n",
       "\n",
       "‚Äòscore(x), score(x) <- value‚Äô: Get or set the ‚Äúscore‚Äù column from the\n",
       "          element metadata.\n",
       "\n",
       "‚Äògranges(x, use.names=FALSE, use.mcols=FALSE)‚Äô: Squeeze the genomic\n",
       "          ranges out of GenomicRanges object ‚Äòx‚Äô and return them in a\n",
       "          GRanges object _parallel_ to ‚Äòx‚Äô (i.e. same length as ‚Äòx‚Äô).\n",
       "          If ‚Äòuse.mcols‚Äô is ‚ÄòTRUE‚Äô, the metadata columns are\n",
       "          propagated.  If ‚Äòx‚Äô is a GenomicRanges derivative with _extra\n",
       "          column slots_, these will be propagated as metadata columns\n",
       "          on the returned GRanges object.\n",
       "\n",
       "_\bC_\bo_\be_\br_\bc_\bi_\bo_\bn:\n",
       "\n",
       "     In the code snippets below, ‚Äòx‚Äô is a GRanges object.\n",
       "\n",
       "‚Äòas(from, \"GRanges\")‚Äô: Creates a GRanges object from a character\n",
       "          vector, a factor, or IntegerRangesList object.\n",
       "\n",
       "          When ‚Äòfrom‚Äô is a character vector (or a factor), each element\n",
       "          in it must represent a genomic range in format\n",
       "          ‚Äòchr1:2501-2800‚Äô (unstranded range) or ‚Äòchr1:2501-2800:+‚Äô\n",
       "          (stranded range).  ‚Äò..‚Äô is also supported as a separator\n",
       "          between the start and end positions. Strand can be ‚Äò+‚Äô, ‚Äò-‚Äô,\n",
       "          ‚Äò*‚Äô, or missing.  The names on ‚Äòfrom‚Äô are propagated to the\n",
       "          returned GRanges object.  See ‚Äòas.character()‚Äô and\n",
       "          ‚Äòas.factor()‚Äô below for the reverse transformations.\n",
       "\n",
       "          Coercing a data.frame or DataFrame into a GRanges object is\n",
       "          also supported. See ‚ÄòmakeGRangesFromDataFrame‚Äô for the\n",
       "          details.\n",
       "\n",
       "‚Äòas(from, \"IntegerRangesList\")‚Äô: Creates a IntegerRangesList object\n",
       "          from a GRanges object. The ‚Äòstrand‚Äô and metadata columns\n",
       "          become _inner_ metadata columns (i.e. metadata columns on the\n",
       "          ranges).  The ‚Äòseqlengths(from)‚Äô, ‚ÄòisCircular(from)‚Äô, and\n",
       "          ‚Äògenome(from)‚Äô vectors become the metadata columns.\n",
       "\n",
       "‚Äòas.character(x, ignore.strand=FALSE)‚Äô: Turn GRanges object ‚Äòx‚Äô into a\n",
       "          character vector where each range in ‚Äòx‚Äô is represented by a\n",
       "          string in format ‚Äòchr1:2501-2800:+‚Äô. If ‚Äòignore.strand‚Äô is\n",
       "          TRUE or if _all_ the ranges in ‚Äòx‚Äô are unstranded (i.e. their\n",
       "          strand is set to ‚Äò*‚Äô), then all the strings in the output are\n",
       "          in format ‚Äòchr1:2501-2800‚Äô.\n",
       "\n",
       "          The names on ‚Äòx‚Äô are propagated to the returned character\n",
       "          vector.  Its metadata (‚Äòmetadata(x)‚Äô) and metadata columns\n",
       "          (‚Äòmcols(x)‚Äô) are ignored.\n",
       "\n",
       "          See ‚Äòas(from, \"GRanges\")‚Äô above for the reverse\n",
       "          transformation.\n",
       "\n",
       "‚Äòas.factor(x)‚Äô: Equivalent to\n",
       "\n",
       "            factor(as.character(x), levels=as.character(sort(unique(x))))\n",
       "          \n",
       "          See ‚Äòas(from, \"GRanges\")‚Äô above for the reverse\n",
       "          transformation.\n",
       "\n",
       "          Note that ‚Äòtable(x)‚Äô is supported on a GRanges object. It is\n",
       "          equivalent to, but much faster than, ‚Äòtable(as.factor(x))‚Äô.\n",
       "\n",
       "‚Äòas.data.frame(x, row.names = NULL, optional = FALSE, ...)‚Äô: Creates a\n",
       "          data.frame with columns ‚Äòseqnames‚Äô (factor), ‚Äòstart‚Äô\n",
       "          (integer), ‚Äòend‚Äô (integer), ‚Äòwidth‚Äô (integer), ‚Äòstrand‚Äô\n",
       "          (factor), as well as the additional metadata columns stored\n",
       "          in ‚Äòmcols(x)‚Äô. Pass an explicit ‚ÄòstringsAsFactors=TRUE/FALSE‚Äô\n",
       "          argument via ‚Äò...‚Äô to override the default conversions for\n",
       "          the metadata columns in ‚Äòmcols(x)‚Äô.\n",
       "\n",
       "‚Äòas(from, \"Grouping\")‚Äô: Creates a ‚ÄòManyToOneGrouping‚Äô object that\n",
       "          groups ‚Äòfrom‚Äô by seqname, strand, start and end (same as the\n",
       "          default sort order). This makes it convenient, for example,\n",
       "          to aggregate a GenomicRanges object by range.\n",
       "\n",
       "     In the code snippets below, ‚Äòx‚Äô is a Seqinfo object.\n",
       "\n",
       "‚Äòas(x, \"GRanges\")‚Äô, ‚Äòas(x, \"GenomicRanges\")‚Äô, ‚Äòas(x,\n",
       "          \"IntegerRangesList\")‚Äô: Turns Seqinfo object ‚Äòx‚Äô (with no ‚ÄòNA‚Äô\n",
       "          lengths) into a GRanges or IntegerRangesList.\n",
       "\n",
       "_\bS_\bu_\bb_\bs_\be_\bt_\bt_\bi_\bn_\bg:\n",
       "\n",
       "     In the code snippets below, ‚Äòx‚Äô is a GRanges object.\n",
       "\n",
       "‚Äòx[i]‚Äô: Return a new GRanges object made of the elements selected by\n",
       "          ‚Äòi‚Äô.\n",
       "\n",
       "‚Äòx[i, j]‚Äô: Like the above, but allow the user to conveniently subset\n",
       "          the metadata columns thru ‚Äòj‚Äô.\n",
       "\n",
       "‚Äòx[i] <- value‚Äô: Replacement version of ‚Äòx[i]‚Äô.\n",
       "\n",
       "‚Äòx$name‚Äô, ‚Äòx$name <- value‚Äô: Shortcuts for ‚Äòmcols(x)$name‚Äô and\n",
       "          ‚Äòmcols(x)$name <- value‚Äô, respectively. Provided as a\n",
       "          convenience, for GRanges objects *only*, and as the result of\n",
       "          strong popular demand.  Note that those methods are not\n",
       "          consistent with the other ‚Äò$‚Äô and ‚Äò$<-‚Äô methods in the\n",
       "          IRanges/GenomicRanges infrastructure, and might confuse some\n",
       "          users by making them believe that a GRanges object can be\n",
       "          manipulated as a data.frame-like object.  Therefore we\n",
       "          recommend using them only interactively, and we discourage\n",
       "          their use in scripts or packages. For the latter, use\n",
       "          ‚Äòmcols(x)$name‚Äô and ‚Äòmcols(x)$name <- value‚Äô, instead of\n",
       "          ‚Äòx$name‚Äô and ‚Äòx$name <- value‚Äô, respectively.\n",
       "\n",
       "     See ‚Äò?`[`‚Äô in the ‚ÄòS4Vectors‚Äô package for more information about\n",
       "     subsetting Vector derivatives and for an important note about the\n",
       "     ‚Äòx[i, j]‚Äô form.\n",
       "\n",
       "     Note that a GRanges object can be used as a subscript to subset a\n",
       "     list-like object that has names on it. In that case, the names on\n",
       "     the list-like object are interpreted as sequence names.  In the\n",
       "     code snippets below, ‚Äòx‚Äô is a list or List object with names on\n",
       "     it, and the subscript ‚Äògr‚Äô is a GRanges object with all its\n",
       "     seqnames being valid ‚Äòx‚Äô names.\n",
       "\n",
       "‚Äòx[gr]‚Äô: Return an object of the same class as ‚Äòx‚Äô and _parallel_ to\n",
       "          ‚Äògr‚Äô. More precisely, it's conceptually doing:\n",
       "\n",
       "            lapply(gr, function(gr1) x[[seqnames(gr1)]][ranges(gr1)])\n",
       "          \n",
       "_\bC_\bo_\bn_\bc_\ba_\bt_\be_\bn_\ba_\bt_\bi_\bo_\bn:\n",
       "\n",
       "‚Äòc(x, ..., ignore.mcols=FALSE)‚Äô: Concatenate GRanges object ‚Äòx‚Äô and the\n",
       "          GRanges objects in ‚Äò...‚Äô together.  See ‚Äò?c‚Äô in the\n",
       "          ‚ÄòS4Vectors‚Äô package for more information about concatenating\n",
       "          Vector derivatives.\n",
       "\n",
       "_\bS_\bp_\bl_\bi_\bt_\bt_\bi_\bn_\bg:\n",
       "\n",
       "‚Äòsplit(x, f, drop=FALSE)‚Äô: Splits GRanges object ‚Äòx‚Äô according to ‚Äòf‚Äô\n",
       "          to create a GRangesList object.  If ‚Äòf‚Äô is a list-like object\n",
       "          then ‚Äòdrop‚Äô is ignored and ‚Äòf‚Äô is treated as if it was\n",
       "          ‚Äòrep(seq_len(length(f)), sapply(f, length))‚Äô, so the returned\n",
       "          object has the same shape as ‚Äòf‚Äô (it also receives the names\n",
       "          of ‚Äòf‚Äô).  Otherwise, if ‚Äòf‚Äô is not a list-like object, empty\n",
       "          list elements are removed from the returned object if ‚Äòdrop‚Äô\n",
       "          is ‚ÄòTRUE‚Äô.\n",
       "\n",
       "_\bD_\bi_\bs_\bp_\bl_\ba_\by_\bi_\bn_\bg:\n",
       "\n",
       "     In the code snippets below, ‚Äòx‚Äô is a GRanges object.\n",
       "\n",
       "‚Äòshow(x)‚Äô: By default the ‚Äòshow‚Äô method displays 5 head and 5 tail\n",
       "          elements. This can be changed by setting the global options\n",
       "          ‚ÄòshowHeadLines‚Äô and ‚ÄòshowTailLines‚Äô. If the object length is\n",
       "          less than (or equal to) the sum of these 2 options plus 1,\n",
       "          then the full object is displayed.  Note that these options\n",
       "          also affect the display of GAlignments and GAlignmentPairs\n",
       "          objects (defined in the ‚ÄòGenomicAlignments‚Äô package), as well\n",
       "          as other objects defined in the ‚ÄòIRanges‚Äô and ‚ÄòBiostrings‚Äô\n",
       "          packages (e.g.  IRanges and DNAStringSet objects).\n",
       "\n",
       "_\bA_\bu_\bt_\bh_\bo_\br(_\bs):\n",
       "\n",
       "     P. Aboyoun and H. Pag√®s\n",
       "\n",
       "_\bS_\be_\be _\bA_\bl_\bs_\bo:\n",
       "\n",
       "        ‚Ä¢ The IRanges class in the ‚ÄòIRanges‚Äô package for storing a set\n",
       "          of _integer ranges_.\n",
       "\n",
       "        ‚Ä¢ The GPos class for representing a set of _genomic positions_\n",
       "          (i.e. _genomic ranges_ of width 1, a.k.a.  _genomic loci_).\n",
       "\n",
       "        ‚Ä¢ ‚ÄòmakeGRangesFromDataFrame‚Äô for making a GRanges object from a\n",
       "          data.frame or DataFrame object.\n",
       "\n",
       "        ‚Ä¢ Seqinfo objects and the ‚Äòseqinfo‚Äô accessor and family in the\n",
       "          ‚ÄòGenomeInfoDb‚Äô package for accessing/modifying information\n",
       "          about the underlying sequences of a GenomicRanges derivative.\n",
       "\n",
       "        ‚Ä¢ GenomicRanges-comparison for comparing and ordering genomic\n",
       "          ranges and/or positions.\n",
       "\n",
       "        ‚Ä¢ findOverlaps-methods for finding overlapping genomic ranges\n",
       "          and/or positions.\n",
       "\n",
       "        ‚Ä¢ intra-range-methods and inter-range-methods for _intra range_\n",
       "          and _inter range_ transformations of GenomicRanges\n",
       "          derivatives.\n",
       "\n",
       "        ‚Ä¢ coverage-methods for computing the coverage of a set of\n",
       "          genomic ranges and/or positions.\n",
       "\n",
       "        ‚Ä¢ setops-methods for set operations on GRanges objects.\n",
       "\n",
       "        ‚Ä¢ nearest-methods for finding the nearest genomic\n",
       "          range/position neighbor.\n",
       "\n",
       "        ‚Ä¢ ‚ÄòabsoluteRanges‚Äô for transforming genomic ranges into\n",
       "          _absolute_ ranges (i.e. into ranges on the sequence obtained\n",
       "          by virtually concatenating all the sequences in a genome).\n",
       "\n",
       "        ‚Ä¢ ‚ÄòtileGenome‚Äô for putting tiles on a genome.\n",
       "\n",
       "        ‚Ä¢ genomicvars for manipulating genomic variables.\n",
       "\n",
       "        ‚Ä¢ GRangesList objects.\n",
       "\n",
       "        ‚Ä¢ Vector, Rle, and DataFrame objects in the ‚ÄòS4Vectors‚Äô\n",
       "          package.\n",
       "\n",
       "_\bE_\bx_\ba_\bm_\bp_\bl_\be_\bs:\n",
       "\n",
       "     showClass(\"GRanges\")  # shows the known subclasses\n",
       "     \n",
       "     ## ---------------------------------------------------------------------\n",
       "     ## CONSTRUCTION\n",
       "     ## ---------------------------------------------------------------------\n",
       "     ## Specifying the bare minimum i.e. seqnames and ranges only. The\n",
       "     ## GRanges object will have no names, no strand information, and no\n",
       "     ## metadata columns:\n",
       "     gr0 <- GRanges(Rle(c(\"chr2\", \"chr2\", \"chr1\", \"chr3\"), c(1, 3, 2, 4)),\n",
       "                    IRanges(1:10, width=10:1))\n",
       "     gr0\n",
       "     \n",
       "     ## Specifying names, strand, metadata columns. They can be set on an\n",
       "     ## existing object:\n",
       "     names(gr0) <- head(letters, 10)\n",
       "     strand(gr0) <- Rle(strand(c(\"-\", \"+\", \"*\", \"+\", \"-\")), c(1, 2, 2, 3, 2))\n",
       "     mcols(gr0)$score <- 1:10\n",
       "     mcols(gr0)$GC <- seq(1, 0, length=10)\n",
       "     gr0\n",
       "     \n",
       "     ## ... or specified at construction time:\n",
       "     gr <- GRanges(Rle(c(\"chr2\", \"chr2\", \"chr1\", \"chr3\"), c(1, 3, 2, 4)),\n",
       "                   IRanges(1:10, width=10:1, names=head(letters, 10)),\n",
       "                   Rle(strand(c(\"-\", \"+\", \"*\", \"+\", \"-\")), c(1, 2, 2, 3, 2)),\n",
       "                   score=1:10, GC=seq(1, 0, length=10))\n",
       "     stopifnot(identical(gr0, gr))\n",
       "     \n",
       "     ## Specifying the seqinfo. It can be set on an existing object:\n",
       "     seqinfo <- Seqinfo(paste0(\"chr\", 1:3), c(1000, 2000, 1500), NA, \"mock1\")\n",
       "     seqinfo(gr0) <- merge(seqinfo(gr0), seqinfo)\n",
       "     seqlevels(gr0) <- seqlevels(seqinfo)\n",
       "     \n",
       "     ## ... or specified at construction time:\n",
       "     gr <- GRanges(Rle(c(\"chr2\", \"chr2\", \"chr1\", \"chr3\"), c(1, 3, 2, 4)),\n",
       "                   IRanges(1:10, width=10:1, names=head(letters, 10)),\n",
       "                   Rle(strand(c(\"-\", \"+\", \"*\", \"+\", \"-\")), c(1, 2, 2, 3, 2)),\n",
       "                   score=1:10, GC=seq(1, 0, length=10),\n",
       "                   seqinfo=seqinfo)\n",
       "     stopifnot(identical(gr0, gr))\n",
       "     \n",
       "     ## ---------------------------------------------------------------------\n",
       "     ## COERCION\n",
       "     ## ---------------------------------------------------------------------\n",
       "     ## From GRanges:\n",
       "     as.character(gr)\n",
       "     as.factor(gr)\n",
       "     as.data.frame(gr)\n",
       "     \n",
       "     ## From character to GRanges:\n",
       "     x1 <- \"chr2:56-125\"\n",
       "     as(x1, \"GRanges\")\n",
       "     as(rep(x1, 4), \"GRanges\")\n",
       "     x2 <- c(A=x1, B=\"chr1:25-30:-\")\n",
       "     as(x2, \"GRanges\")\n",
       "     \n",
       "     ## From data.frame to GRanges:\n",
       "     df <- data.frame(chrom=\"chr2\", start=11:15, end=20:24)\n",
       "     gr3 <- as(df, \"GRanges\")\n",
       "     \n",
       "     ## Alternatively, coercion to GRanges can be done by just calling the\n",
       "     ## GRanges() constructor on the object to coerce:\n",
       "     gr1 <- GRanges(x1)  # same as as(x1, \"GRanges\")\n",
       "     gr2 <- GRanges(x2)  # same as as(x2, \"GRanges\")\n",
       "     gr3 <- GRanges(df)  # same as as(df, \"GRanges\")\n",
       "     \n",
       "     ## Sanity checks:\n",
       "     stopifnot(identical(as(x1, \"GRanges\"), gr1))\n",
       "     stopifnot(identical(as(x2, \"GRanges\"), gr2))\n",
       "     stopifnot(identical(as(df, \"GRanges\"), gr3))\n",
       "     \n",
       "     ## ---------------------------------------------------------------------\n",
       "     ## SUMMARIZING ELEMENTS\n",
       "     ## ---------------------------------------------------------------------\n",
       "     table(seqnames(gr))\n",
       "     table(strand(gr))\n",
       "     sum(width(gr))\n",
       "     table(gr)\n",
       "     summary(mcols(gr)[,\"score\"])\n",
       "     \n",
       "     ## The number of lines displayed in the 'show' method are controlled\n",
       "     ## with two global options:\n",
       "     longGR  <- sample(gr, 25, replace=TRUE)\n",
       "     longGR\n",
       "     options(showHeadLines=7)\n",
       "     options(showTailLines=2)\n",
       "     longGR\n",
       "     \n",
       "     ## Revert to default values\n",
       "     options(showHeadLines=NULL)\n",
       "     options(showTailLines=NULL)\n",
       "     \n",
       "     ## ---------------------------------------------------------------------\n",
       "     ## INVERTING THE STRAND\n",
       "     ## ---------------------------------------------------------------------\n",
       "     invertStrand(gr)\n",
       "     \n",
       "     ## ---------------------------------------------------------------------\n",
       "     ## RENAMING THE UNDERLYING SEQUENCES\n",
       "     ## ---------------------------------------------------------------------\n",
       "     seqlevels(gr)\n",
       "     seqlevels(gr) <- sub(\"chr\", \"Chrom\", seqlevels(gr))\n",
       "     gr\n",
       "     seqlevels(gr) <- sub(\"Chrom\", \"chr\", seqlevels(gr)) # revert\n",
       "     \n",
       "     ## ---------------------------------------------------------------------\n",
       "     ## COMBINING OBJECTS\n",
       "     ## ---------------------------------------------------------------------\n",
       "     gr2 <- GRanges(seqnames=Rle(c('chr1', 'chr2', 'chr3'), c(3, 3, 4)),\n",
       "                    IRanges(1:10, width=5),\n",
       "                    strand='-',\n",
       "                    score=101:110, GC=runif(10),\n",
       "                    seqinfo=seqinfo)\n",
       "     gr3 <- GRanges(seqnames=Rle(c('chr1', 'chr2', 'chr3'), c(3, 4, 3)),\n",
       "                    IRanges(101:110, width=10),\n",
       "                    strand='-',\n",
       "                    score=21:30,\n",
       "                    seqinfo=seqinfo)\n",
       "     some.gr <- c(gr, gr2)\n",
       "     \n",
       "     c(gr, gr2, gr3)\n",
       "     c(gr, gr2, gr3, ignore.mcols=TRUE)\n",
       "     \n",
       "     ## ---------------------------------------------------------------------\n",
       "     ## USING A GRANGES OBJECT AS A SUBSCRIPT TO SUBSET ANOTHER OBJECT\n",
       "     ## ---------------------------------------------------------------------\n",
       "     ## Subsetting *by* a GRanges subscript is supported only if the object\n",
       "     ## to subset is a named list-like object:\n",
       "     x <- RleList(chr1=101:120, chr2=2:-8, chr3=31:40)\n",
       "     x[gr]\n",
       "     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#cell 22\n",
    "?GRanges"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- G√©n√©rez des objets de type `GRanges` avec la function `makeGRangesFromDataFrame` du paquet **GenomicRanges** pour chacun des deux dataframes en ne gardant que les colonnes 2 √† 7 pour les SNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 62 ranges and 1 metadata column:\n",
       "       seqnames            ranges strand |        name\n",
       "          <Rle>         <IRanges>  <Rle> | <character>\n",
       "   [1]    chr17 43024272-43024273      + |   rs7224581\n",
       "   [2]    chr17 43024274-43024275      + |  rs71367982\n",
       "   [3]    chr17 43024279-43024280      + |   rs7222911\n",
       "   [4]    chr17 43024293-43024294      + |   rs7224048\n",
       "   [5]    chr17 43024296-43024297      + |   rs7224049\n",
       "   ...      ...               ...    ... .         ...\n",
       "  [58]    chr17 43044804-43044809      + | rs757144777\n",
       "  [59]    chr17          43044804      + | rs749410065\n",
       "  [60]    chr17 43045256-43045257      - |   rs8176318\n",
       "  [61]    chr17          43045689      + | rs138690298\n",
       "  [62]    chr17 43045690-43045691      + | rs201196020\n",
       "  -------\n",
       "  seqinfo: 1 sequence from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Formal class 'GRanges' [package \"GenomicRanges\"] with 7 slots\n",
      "  ..@ seqnames       :Formal class 'Rle' [package \"S4Vectors\"] with 4 slots\n",
      "  .. .. ..@ values         : Factor w/ 1 level \"chr17\": 1\n",
      "  .. .. ..@ lengths        : int 62\n",
      "  .. .. ..@ elementMetadata: NULL\n",
      "  .. .. ..@ metadata       : list()\n",
      "  ..@ ranges         :Formal class 'IRanges' [package \"IRanges\"] with 6 slots\n",
      "  .. .. ..@ start          : int [1:62] 43024272 43024274 43024279 43024293 43024296 43024629 43024670 43024813 43025042 43025790 ...\n",
      "  .. .. ..@ width          : int [1:62] 2 2 2 2 2 2 2 2 2 2 ...\n",
      "  .. .. ..@ NAMES          : NULL\n",
      "  .. .. ..@ elementType    : chr \"ANY\"\n",
      "  .. .. ..@ elementMetadata: NULL\n",
      "  .. .. ..@ metadata       : list()\n",
      "  ..@ strand         :Formal class 'Rle' [package \"S4Vectors\"] with 4 slots\n",
      "  .. .. ..@ values         : Factor w/ 3 levels \"+\",\"-\",\"*\": 1 2 1 2 1 2 1 2 1 2 ...\n",
      "  .. .. ..@ lengths        : int [1:11] 6 2 3 2 39 2 1 1 3 1 ...\n",
      "  .. .. ..@ elementMetadata: NULL\n",
      "  .. .. ..@ metadata       : list()\n",
      "  ..@ seqinfo        :Formal class 'Seqinfo' [package \"GenomeInfoDb\"] with 4 slots\n",
      "  .. .. ..@ seqnames   : chr \"chr17\"\n",
      "  .. .. ..@ seqlengths : int NA\n",
      "  .. .. ..@ is_circular: logi NA\n",
      "  .. .. ..@ genome     : chr NA\n",
      "  ..@ elementMetadata:Formal class 'DFrame' [package \"S4Vectors\"] with 6 slots\n",
      "  .. .. ..@ rownames       : NULL\n",
      "  .. .. ..@ nrows          : int 62\n",
      "  .. .. ..@ listData       :List of 1\n",
      "  .. .. .. ..$ name: chr [1:62] \"rs7224581\" \"rs71367982\" \"rs7222911\" \"rs7224048\" ...\n",
      "  .. .. ..@ elementType    : chr \"ANY\"\n",
      "  .. .. ..@ elementMetadata: NULL\n",
      "  .. .. ..@ metadata       : list()\n",
      "  ..@ elementType    : chr \"ANY\"\n",
      "  ..@ metadata       : list()\n"
     ]
    }
   ],
   "source": [
    "#cell 23\n",
    "snps_GR <- makeGRangesFromDataFrame(snps[,c(2:5,7)], keep.extra.columns = TRUE)\n",
    "snps_GR\n",
    "str(snps_GR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 30 ranges and 1 metadata column:\n",
       "       seqnames            ranges strand |          id\n",
       "          <Rle>         <IRanges>  <Rle> | <character>\n",
       "   [1]    chr17 43023996-43024058      * |     chr17.1\n",
       "   [2]    chr17 43024283-43024633      * |     chr17.2\n",
       "   [3]    chr17 43024658-43025383      * |     chr17.3\n",
       "   [4]    chr17 43025608-43025833      * |     chr17.4\n",
       "   [5]    chr17 43026183-43026458      * |     chr17.5\n",
       "   ...      ...               ...    ... .         ...\n",
       "  [26]    chr17 43041533-43042033      * |    chr17.26\n",
       "  [27]    chr17 43042308-43042858      * |    chr17.27\n",
       "  [28]    chr17 43042908-43043508      * |    chr17.28\n",
       "  [29]    chr17 43043708-43044558      * |    chr17.29\n",
       "  [30]    chr17 43044783-43045708      * |    chr17.30\n",
       "  -------\n",
       "  seqinfo: 1 sequence from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Formal class 'GRanges' [package \"GenomicRanges\"] with 7 slots\n",
      "  ..@ seqnames       :Formal class 'Rle' [package \"S4Vectors\"] with 4 slots\n",
      "  .. .. ..@ values         : Factor w/ 1 level \"chr17\": 1\n",
      "  .. .. ..@ lengths        : int 30\n",
      "  .. .. ..@ elementMetadata: NULL\n",
      "  .. .. ..@ metadata       : list()\n",
      "  ..@ ranges         :Formal class 'IRanges' [package \"IRanges\"] with 6 slots\n",
      "  .. .. ..@ start          : int [1:30] 43023996 43024283 43024658 43025608 43026183 43027058 43027833 43028208 43029008 43029258 ...\n",
      "  .. .. ..@ width          : int [1:30] 63 351 726 226 276 201 351 201 226 251 ...\n",
      "  .. .. ..@ NAMES          : NULL\n",
      "  .. .. ..@ elementType    : chr \"ANY\"\n",
      "  .. .. ..@ elementMetadata: NULL\n",
      "  .. .. ..@ metadata       : list()\n",
      "  ..@ strand         :Formal class 'Rle' [package \"S4Vectors\"] with 4 slots\n",
      "  .. .. ..@ values         : Factor w/ 3 levels \"+\",\"-\",\"*\": 3\n",
      "  .. .. ..@ lengths        : int 30\n",
      "  .. .. ..@ elementMetadata: NULL\n",
      "  .. .. ..@ metadata       : list()\n",
      "  ..@ seqinfo        :Formal class 'Seqinfo' [package \"GenomeInfoDb\"] with 4 slots\n",
      "  .. .. ..@ seqnames   : chr \"chr17\"\n",
      "  .. .. ..@ seqlengths : int NA\n",
      "  .. .. ..@ is_circular: logi NA\n",
      "  .. .. ..@ genome     : chr NA\n",
      "  ..@ elementMetadata:Formal class 'DFrame' [package \"S4Vectors\"] with 6 slots\n",
      "  .. .. ..@ rownames       : NULL\n",
      "  .. .. ..@ nrows          : int 30\n",
      "  .. .. ..@ listData       :List of 1\n",
      "  .. .. .. ..$ id: chr [1:30] \"chr17.1\" \"chr17.2\" \"chr17.3\" \"chr17.4\" ...\n",
      "  .. .. ..@ elementType    : chr \"ANY\"\n",
      "  .. .. ..@ elementMetadata: NULL\n",
      "  .. .. ..@ metadata       : list()\n",
      "  ..@ elementType    : chr \"ANY\"\n",
      "  ..@ metadata       : list()\n"
     ]
    }
   ],
   "source": [
    "#cell 24\n",
    "h3k27ac_GR <- makeGRangesFromDataFrame(h3k27ac, keep.extra.columns = TRUE)\n",
    "h3k27ac_GR\n",
    "str(h3k27ac_GR)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Si vous inspectez les deux objets, ils ont tous les deux 7 slots comme tout objet de type GenomicRanges"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Identifiez abvec la fonction `findOverlaps()` les chevauchements entre les deux datasets, c'est √† dire entre les SNPs et les sites d'ac√©tylation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Hits object with 35 hits and 0 metadata columns:\n",
       "       queryHits subjectHits\n",
       "       <integer>   <integer>\n",
       "   [1]         2           4\n",
       "   [2]         2           5\n",
       "   [3]         2           6\n",
       "   [4]         3           7\n",
       "   [5]         3           8\n",
       "   ...       ...         ...\n",
       "  [31]        30          58\n",
       "  [32]        30          59\n",
       "  [33]        30          60\n",
       "  [34]        30          61\n",
       "  [35]        30          62\n",
       "  -------\n",
       "  queryLength: 30 / subjectLength: 62"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#cell 25\n",
    "# Find overlaps between SNPs and acetylation sites\n",
    "overlaps <- findOverlaps(h3k27ac_GR, snps_GR)\n",
    "overlaps"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "La commande suivante vous donne le nombre d'overlaps par √©l√©ment, ici combien de SNPs sont trouv√©s pour chaque site d'acetylation (de 0 √† 6 selon le site)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " [1] 0 3 3 2 0 0 2 0 0 1 2 0 1 1 0 2 0 1 1 1 2 2 0 0 0 1 0 1 3 6\n"
     ]
    }
   ],
   "source": [
    "#cell 26\n",
    "print(countOverlaps(h3k27ac_GR, snps_GR))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Vous pouvez g√©n√©rer et sauvergarder  un sous-jeu des 18 sites d'actylation contenant au moins un SNP avec la commande suivante:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Formal class 'GRanges' [package \"GenomicRanges\"] with 7 slots\n",
      "  ..@ seqnames       :Formal class 'Rle' [package \"S4Vectors\"] with 4 slots\n",
      "  .. .. ..@ values         : Factor w/ 1 level \"chr17\": 1\n",
      "  .. .. ..@ lengths        : int 18\n",
      "  .. .. ..@ elementMetadata: NULL\n",
      "  .. .. ..@ metadata       : list()\n",
      "  ..@ ranges         :Formal class 'IRanges' [package \"IRanges\"] with 6 slots\n",
      "  .. .. ..@ start          : int [1:18] 43024283 43024658 43025608 43027833 43029258 43029733 43033283 43034208 43035633 43037183 ...\n",
      "  .. .. ..@ width          : int [1:18] 351 726 226 351 251 226 901 626 401 226 ...\n",
      "  .. .. ..@ NAMES          : NULL\n",
      "  .. .. ..@ elementType    : chr \"ANY\"\n",
      "  .. .. ..@ elementMetadata: NULL\n",
      "  .. .. ..@ metadata       : list()\n",
      "  ..@ strand         :Formal class 'Rle' [package \"S4Vectors\"] with 4 slots\n",
      "  .. .. ..@ values         : Factor w/ 3 levels \"+\",\"-\",\"*\": 3\n",
      "  .. .. ..@ lengths        : int 18\n",
      "  .. .. ..@ elementMetadata: NULL\n",
      "  .. .. ..@ metadata       : list()\n",
      "  ..@ seqinfo        :Formal class 'Seqinfo' [package \"GenomeInfoDb\"] with 4 slots\n",
      "  .. .. ..@ seqnames   : chr \"chr17\"\n",
      "  .. .. ..@ seqlengths : int NA\n",
      "  .. .. ..@ is_circular: logi NA\n",
      "  .. .. ..@ genome     : chr NA\n",
      "  ..@ elementMetadata:Formal class 'DFrame' [package \"S4Vectors\"] with 6 slots\n",
      "  .. .. ..@ rownames       : NULL\n",
      "  .. .. ..@ nrows          : int 18\n",
      "  .. .. ..@ listData       :List of 1\n",
      "  .. .. .. ..$ id: chr [1:18] \"chr17.2\" \"chr17.3\" \"chr17.4\" \"chr17.7\" ...\n",
      "  .. .. ..@ elementType    : chr \"ANY\"\n",
      "  .. .. ..@ elementMetadata: NULL\n",
      "  .. .. ..@ metadata       : list()\n",
      "  ..@ elementType    : chr \"ANY\"\n",
      "  ..@ metadata       : list()\n"
     ]
    }
   ],
   "source": [
    "#cell 27\n",
    "h3k27ac_SNPs_GR <- subsetByOverlaps(h3k27ac_GR, snps_GR)\n",
    "str(h3k27ac_SNPs_GR)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.3. Utilisation du paquet **biomaRt**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3.3.1. Cherchez de la documentation sur ce paquet\n",
    "\n",
    "- Consultez la documentation de BioMart: https://bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/biomaRt.html"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Quelles sont les bases de donn√©es interrogeables dans la base de donnees ensembl avec biomaRt ?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 4 √ó 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>biomart</th><th scope=col>version</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>ENSEMBL_MART_ENSEMBL</td><td>Ensembl Genes 104     </td></tr>\n",
       "\t<tr><td>ENSEMBL_MART_MOUSE  </td><td>Mouse strains 104     </td></tr>\n",
       "\t<tr><td>ENSEMBL_MART_SNP    </td><td>Ensembl Variation 104 </td></tr>\n",
       "\t<tr><td>ENSEMBL_MART_FUNCGEN</td><td>Ensembl Regulation 104</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 4 √ó 2\n",
       "\\begin{tabular}{ll}\n",
       " biomart & version\\\\\n",
       " <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t ENSEMBL\\_MART\\_ENSEMBL & Ensembl Genes 104     \\\\\n",
       "\t ENSEMBL\\_MART\\_MOUSE   & Mouse strains 104     \\\\\n",
       "\t ENSEMBL\\_MART\\_SNP     & Ensembl Variation 104 \\\\\n",
       "\t ENSEMBL\\_MART\\_FUNCGEN & Ensembl Regulation 104\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 4 √ó 2\n",
       "\n",
       "| biomart &lt;chr&gt; | version &lt;chr&gt; |\n",
       "|---|---|\n",
       "| ENSEMBL_MART_ENSEMBL | Ensembl Genes 104      |\n",
       "| ENSEMBL_MART_MOUSE   | Mouse strains 104      |\n",
       "| ENSEMBL_MART_SNP     | Ensembl Variation 104  |\n",
       "| ENSEMBL_MART_FUNCGEN | Ensembl Regulation 104 |\n",
       "\n"
      ],
      "text/plain": [
       "  biomart              version               \n",
       "1 ENSEMBL_MART_ENSEMBL Ensembl Genes 104     \n",
       "2 ENSEMBL_MART_MOUSE   Mouse strains 104     \n",
       "3 ENSEMBL_MART_SNP     Ensembl Variation 104 \n",
       "4 ENSEMBL_MART_FUNCGEN Ensembl Regulation 104"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#cell 28\n",
    "listMarts(host = \"www.ensembl.org\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- S√©lectionner √† pr√©sent le dataset Homo Sapiens de la base de donnees Ensembl avec la fonction `useDataset()` ou directement `useMart()` :"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "#cell 29\n",
    "ensembl <- useMart(\"ENSEMBL_MART_ENSEMBL\", host = \"www.ensembl.org\", dataset = \"hsapiens_gene_ensembl\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Combien y a-t- il de bases de datasets interrogeables dans cette base de donnees ?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "202"
      ],
      "text/latex": [
       "202"
      ],
      "text/markdown": [
       "202"
      ],
      "text/plain": [
       "[1] 202"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#cell 30\n",
    "dim(listDatasets(ensembl))[1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Combien y a-t- il de champs interrogeables dans la base de donnees ensembl Homo Sapiens ?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "3164"
      ],
      "text/latex": [
       "3164"
      ],
      "text/markdown": [
       "3164"
      ],
      "text/plain": [
       "[1] 3164"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#cell 31\n",
    "nrow(listAttributes(ensembl))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3.3.2. Identifier les genes √† +/- 20 kb des sites d'actelylation contenant au moins un SNP avec la fonction `getBM` du paquet `biomaRt`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "#cell 32\n",
    "info <- data.frame(h3k27ac_SNPs_GR@seqnames@values, h3k27ac_SNPs_GR@ranges@start,\n",
    "                   h3k27ac_SNPs_GR@ranges@start + h3k27ac_SNPs_GR@ranges@width)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "'data.frame':\t18 obs. of  3 variables:\n",
      " $ h3k27ac_SNPs_GR.seqnames.values                            : Factor w/ 1 level \"chr17\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      " $ h3k27ac_SNPs_GR.ranges.start                               : int  43024283 43024658 43025608 43027833 43029258 43029733 43033283 43034208 43035633 43037183 ...\n",
      " $ h3k27ac_SNPs_GR.ranges.start...h3k27ac_SNPs_GR.ranges.width: int  43024634 43025384 43025834 43028184 43029509 43029959 43034184 43034834 43036034 43037409 ...\n"
     ]
    }
   ],
   "source": [
    "#cell 33\n",
    "str(info)\n",
    "names(info) <- c(\"chr\", \"start\", \"end\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 35 √ó 8</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>site</th><th scope=col>hgnc_symbol</th><th scope=col>description</th><th scope=col>ensembl_gene_id</th><th scope=col>chromosome_name</th><th scope=col>start_position</th><th scope=col>end_position</th><th scope=col>strand</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr17:43024283-43024634</td><td>IFI35</td><td>interferon induced protein 35 [Source:HGNC Symbol;Acc:HGNC:5399]</td><td>ENSG00000068079</td><td>17</td><td>43006740</td><td>43014456</td><td> 1</td></tr>\n",
       "\t<tr><td>chr17:43024283-43024634</td><td>VAT1 </td><td>vesicle amine transport 1 [Source:HGNC Symbol;Acc:HGNC:16919]   </td><td>ENSG00000108828</td><td>17</td><td>43014607</td><td>43025123</td><td>-1</td></tr>\n",
       "\t<tr><td>chr17:43024283-43024634</td><td>RND2 </td><td>Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]         </td><td>ENSG00000108830</td><td>17</td><td>43025231</td><td>43032041</td><td> 1</td></tr>\n",
       "\t<tr><td>chr17:43024658-43025384</td><td>VAT1 </td><td>vesicle amine transport 1 [Source:HGNC Symbol;Acc:HGNC:16919]   </td><td>ENSG00000108828</td><td>17</td><td>43014607</td><td>43025123</td><td>-1</td></tr>\n",
       "\t<tr><td>chr17:43024658-43025384</td><td>RND2 </td><td>Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]         </td><td>ENSG00000108830</td><td>17</td><td>43025231</td><td>43032041</td><td> 1</td></tr>\n",
       "\t<tr><td>chr17:43025608-43025834</td><td>VAT1 </td><td>vesicle amine transport 1 [Source:HGNC Symbol;Acc:HGNC:16919]   </td><td>ENSG00000108828</td><td>17</td><td>43014607</td><td>43025123</td><td>-1</td></tr>\n",
       "\t<tr><td>chr17:43025608-43025834</td><td>RND2 </td><td>Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]         </td><td>ENSG00000108830</td><td>17</td><td>43025231</td><td>43032041</td><td> 1</td></tr>\n",
       "\t<tr><td>chr17:43027833-43028184</td><td>VAT1 </td><td>vesicle amine transport 1 [Source:HGNC Symbol;Acc:HGNC:16919]   </td><td>ENSG00000108828</td><td>17</td><td>43014607</td><td>43025123</td><td>-1</td></tr>\n",
       "\t<tr><td>chr17:43027833-43028184</td><td>RND2 </td><td>Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]         </td><td>ENSG00000108830</td><td>17</td><td>43025231</td><td>43032041</td><td> 1</td></tr>\n",
       "\t<tr><td>chr17:43029258-43029509</td><td>VAT1 </td><td>vesicle amine transport 1 [Source:HGNC Symbol;Acc:HGNC:16919]   </td><td>ENSG00000108828</td><td>17</td><td>43014607</td><td>43025123</td><td>-1</td></tr>\n",
       "\t<tr><td>chr17:43029258-43029509</td><td>RND2 </td><td>Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]         </td><td>ENSG00000108830</td><td>17</td><td>43025231</td><td>43032041</td><td> 1</td></tr>\n",
       "\t<tr><td>chr17:43029733-43029959</td><td>VAT1 </td><td>vesicle amine transport 1 [Source:HGNC Symbol;Acc:HGNC:16919]   </td><td>ENSG00000108828</td><td>17</td><td>43014607</td><td>43025123</td><td>-1</td></tr>\n",
       "\t<tr><td>chr17:43029733-43029959</td><td>RND2 </td><td>Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]         </td><td>ENSG00000108830</td><td>17</td><td>43025231</td><td>43032041</td><td> 1</td></tr>\n",
       "\t<tr><td>chr17:43033283-43034184</td><td>VAT1 </td><td>vesicle amine transport 1 [Source:HGNC Symbol;Acc:HGNC:16919]   </td><td>ENSG00000108828</td><td>17</td><td>43014607</td><td>43025123</td><td>-1</td></tr>\n",
       "\t<tr><td>chr17:43033283-43034184</td><td>RND2 </td><td>Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]         </td><td>ENSG00000108830</td><td>17</td><td>43025231</td><td>43032041</td><td> 1</td></tr>\n",
       "\t<tr><td>chr17:43034208-43034834</td><td>VAT1 </td><td>vesicle amine transport 1 [Source:HGNC Symbol;Acc:HGNC:16919]   </td><td>ENSG00000108828</td><td>17</td><td>43014607</td><td>43025123</td><td>-1</td></tr>\n",
       "\t<tr><td>chr17:43034208-43034834</td><td>RND2 </td><td>Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]         </td><td>ENSG00000108830</td><td>17</td><td>43025231</td><td>43032041</td><td> 1</td></tr>\n",
       "\t<tr><td>chr17:43034208-43034834</td><td>BRCA1</td><td>BRCA1 DNA repair associated [Source:HGNC Symbol;Acc:HGNC:1100]  </td><td>ENSG00000012048</td><td>17</td><td>43044295</td><td>43170245</td><td>-1</td></tr>\n",
       "\t<tr><td>chr17:43035633-43036034</td><td>RND2 </td><td>Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]         </td><td>ENSG00000108830</td><td>17</td><td>43025231</td><td>43032041</td><td> 1</td></tr>\n",
       "\t<tr><td>chr17:43035633-43036034</td><td>BRCA1</td><td>BRCA1 DNA repair associated [Source:HGNC Symbol;Acc:HGNC:1100]  </td><td>ENSG00000012048</td><td>17</td><td>43044295</td><td>43170245</td><td>-1</td></tr>\n",
       "\t<tr><td>chr17:43037183-43037409</td><td>RND2 </td><td>Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]         </td><td>ENSG00000108830</td><td>17</td><td>43025231</td><td>43032041</td><td> 1</td></tr>\n",
       "\t<tr><td>chr17:43037183-43037409</td><td>BRCA1</td><td>BRCA1 DNA repair associated [Source:HGNC Symbol;Acc:HGNC:1100]  </td><td>ENSG00000012048</td><td>17</td><td>43044295</td><td>43170245</td><td>-1</td></tr>\n",
       "\t<tr><td>chr17:43037658-43038184</td><td>RND2 </td><td>Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]         </td><td>ENSG00000108830</td><td>17</td><td>43025231</td><td>43032041</td><td> 1</td></tr>\n",
       "\t<tr><td>chr17:43037658-43038184</td><td>BRCA1</td><td>BRCA1 DNA repair associated [Source:HGNC Symbol;Acc:HGNC:1100]  </td><td>ENSG00000012048</td><td>17</td><td>43044295</td><td>43170245</td><td>-1</td></tr>\n",
       "\t<tr><td>chr17:43038208-43038434</td><td>RND2 </td><td>Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]         </td><td>ENSG00000108830</td><td>17</td><td>43025231</td><td>43032041</td><td> 1</td></tr>\n",
       "\t<tr><td>chr17:43038208-43038434</td><td>BRCA1</td><td>BRCA1 DNA repair associated [Source:HGNC Symbol;Acc:HGNC:1100]  </td><td>ENSG00000012048</td><td>17</td><td>43044295</td><td>43170245</td><td>-1</td></tr>\n",
       "\t<tr><td>chr17:43038558-43038759</td><td>RND2 </td><td>Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]         </td><td>ENSG00000108830</td><td>17</td><td>43025231</td><td>43032041</td><td> 1</td></tr>\n",
       "\t<tr><td>chr17:43038558-43038759</td><td>BRCA1</td><td>BRCA1 DNA repair associated [Source:HGNC Symbol;Acc:HGNC:1100]  </td><td>ENSG00000012048</td><td>17</td><td>43044295</td><td>43170245</td><td>-1</td></tr>\n",
       "\t<tr><td>chr17:43038933-43039159</td><td>RND2 </td><td>Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]         </td><td>ENSG00000108830</td><td>17</td><td>43025231</td><td>43032041</td><td> 1</td></tr>\n",
       "\t<tr><td>chr17:43038933-43039159</td><td>BRCA1</td><td>BRCA1 DNA repair associated [Source:HGNC Symbol;Acc:HGNC:1100]  </td><td>ENSG00000012048</td><td>17</td><td>43044295</td><td>43170245</td><td>-1</td></tr>\n",
       "\t<tr><td>chr17:43041533-43042034</td><td>RND2 </td><td>Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]         </td><td>ENSG00000108830</td><td>17</td><td>43025231</td><td>43032041</td><td> 1</td></tr>\n",
       "\t<tr><td>chr17:43041533-43042034</td><td>BRCA1</td><td>BRCA1 DNA repair associated [Source:HGNC Symbol;Acc:HGNC:1100]  </td><td>ENSG00000012048</td><td>17</td><td>43044295</td><td>43170245</td><td>-1</td></tr>\n",
       "\t<tr><td>chr17:43042908-43043509</td><td>BRCA1</td><td>BRCA1 DNA repair associated [Source:HGNC Symbol;Acc:HGNC:1100]  </td><td>ENSG00000012048</td><td>17</td><td>43044295</td><td>43170245</td><td>-1</td></tr>\n",
       "\t<tr><td>chr17:43043708-43044559</td><td>BRCA1</td><td>BRCA1 DNA repair associated [Source:HGNC Symbol;Acc:HGNC:1100]  </td><td>ENSG00000012048</td><td>17</td><td>43044295</td><td>43170245</td><td>-1</td></tr>\n",
       "\t<tr><td>chr17:43044783-43045709</td><td>BRCA1</td><td>BRCA1 DNA repair associated [Source:HGNC Symbol;Acc:HGNC:1100]  </td><td>ENSG00000012048</td><td>17</td><td>43044295</td><td>43170245</td><td>-1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 35 √ó 8\n",
       "\\begin{tabular}{llllllll}\n",
       " site & hgnc\\_symbol & description & ensembl\\_gene\\_id & chromosome\\_name & start\\_position & end\\_position & strand\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <int> & <int> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t chr17:43024283-43024634 & IFI35 & interferon induced protein 35 {[}Source:HGNC Symbol;Acc:HGNC:5399{]} & ENSG00000068079 & 17 & 43006740 & 43014456 &  1\\\\\n",
       "\t chr17:43024283-43024634 & VAT1  & vesicle amine transport 1 {[}Source:HGNC Symbol;Acc:HGNC:16919{]}    & ENSG00000108828 & 17 & 43014607 & 43025123 & -1\\\\\n",
       "\t chr17:43024283-43024634 & RND2  & Rho family GTPase 2 {[}Source:HGNC Symbol;Acc:HGNC:18315{]}          & ENSG00000108830 & 17 & 43025231 & 43032041 &  1\\\\\n",
       "\t chr17:43024658-43025384 & VAT1  & vesicle amine transport 1 {[}Source:HGNC Symbol;Acc:HGNC:16919{]}    & ENSG00000108828 & 17 & 43014607 & 43025123 & -1\\\\\n",
       "\t chr17:43024658-43025384 & RND2  & Rho family GTPase 2 {[}Source:HGNC Symbol;Acc:HGNC:18315{]}          & ENSG00000108830 & 17 & 43025231 & 43032041 &  1\\\\\n",
       "\t chr17:43025608-43025834 & VAT1  & vesicle amine transport 1 {[}Source:HGNC Symbol;Acc:HGNC:16919{]}    & ENSG00000108828 & 17 & 43014607 & 43025123 & -1\\\\\n",
       "\t chr17:43025608-43025834 & RND2  & Rho family GTPase 2 {[}Source:HGNC Symbol;Acc:HGNC:18315{]}          & ENSG00000108830 & 17 & 43025231 & 43032041 &  1\\\\\n",
       "\t chr17:43027833-43028184 & VAT1  & vesicle amine transport 1 {[}Source:HGNC Symbol;Acc:HGNC:16919{]}    & ENSG00000108828 & 17 & 43014607 & 43025123 & -1\\\\\n",
       "\t chr17:43027833-43028184 & RND2  & Rho family GTPase 2 {[}Source:HGNC Symbol;Acc:HGNC:18315{]}          & ENSG00000108830 & 17 & 43025231 & 43032041 &  1\\\\\n",
       "\t chr17:43029258-43029509 & VAT1  & vesicle amine transport 1 {[}Source:HGNC Symbol;Acc:HGNC:16919{]}    & ENSG00000108828 & 17 & 43014607 & 43025123 & -1\\\\\n",
       "\t chr17:43029258-43029509 & RND2  & Rho family GTPase 2 {[}Source:HGNC Symbol;Acc:HGNC:18315{]}          & ENSG00000108830 & 17 & 43025231 & 43032041 &  1\\\\\n",
       "\t chr17:43029733-43029959 & VAT1  & vesicle amine transport 1 {[}Source:HGNC Symbol;Acc:HGNC:16919{]}    & ENSG00000108828 & 17 & 43014607 & 43025123 & -1\\\\\n",
       "\t chr17:43029733-43029959 & RND2  & Rho family GTPase 2 {[}Source:HGNC Symbol;Acc:HGNC:18315{]}          & ENSG00000108830 & 17 & 43025231 & 43032041 &  1\\\\\n",
       "\t chr17:43033283-43034184 & VAT1  & vesicle amine transport 1 {[}Source:HGNC Symbol;Acc:HGNC:16919{]}    & ENSG00000108828 & 17 & 43014607 & 43025123 & -1\\\\\n",
       "\t chr17:43033283-43034184 & RND2  & Rho family GTPase 2 {[}Source:HGNC Symbol;Acc:HGNC:18315{]}          & ENSG00000108830 & 17 & 43025231 & 43032041 &  1\\\\\n",
       "\t chr17:43034208-43034834 & VAT1  & vesicle amine transport 1 {[}Source:HGNC Symbol;Acc:HGNC:16919{]}    & ENSG00000108828 & 17 & 43014607 & 43025123 & -1\\\\\n",
       "\t chr17:43034208-43034834 & RND2  & Rho family GTPase 2 {[}Source:HGNC Symbol;Acc:HGNC:18315{]}          & ENSG00000108830 & 17 & 43025231 & 43032041 &  1\\\\\n",
       "\t chr17:43034208-43034834 & BRCA1 & BRCA1 DNA repair associated {[}Source:HGNC Symbol;Acc:HGNC:1100{]}   & ENSG00000012048 & 17 & 43044295 & 43170245 & -1\\\\\n",
       "\t chr17:43035633-43036034 & RND2  & Rho family GTPase 2 {[}Source:HGNC Symbol;Acc:HGNC:18315{]}          & ENSG00000108830 & 17 & 43025231 & 43032041 &  1\\\\\n",
       "\t chr17:43035633-43036034 & BRCA1 & BRCA1 DNA repair associated {[}Source:HGNC Symbol;Acc:HGNC:1100{]}   & ENSG00000012048 & 17 & 43044295 & 43170245 & -1\\\\\n",
       "\t chr17:43037183-43037409 & RND2  & Rho family GTPase 2 {[}Source:HGNC Symbol;Acc:HGNC:18315{]}          & ENSG00000108830 & 17 & 43025231 & 43032041 &  1\\\\\n",
       "\t chr17:43037183-43037409 & BRCA1 & BRCA1 DNA repair associated {[}Source:HGNC Symbol;Acc:HGNC:1100{]}   & ENSG00000012048 & 17 & 43044295 & 43170245 & -1\\\\\n",
       "\t chr17:43037658-43038184 & RND2  & Rho family GTPase 2 {[}Source:HGNC Symbol;Acc:HGNC:18315{]}          & ENSG00000108830 & 17 & 43025231 & 43032041 &  1\\\\\n",
       "\t chr17:43037658-43038184 & BRCA1 & BRCA1 DNA repair associated {[}Source:HGNC Symbol;Acc:HGNC:1100{]}   & ENSG00000012048 & 17 & 43044295 & 43170245 & -1\\\\\n",
       "\t chr17:43038208-43038434 & RND2  & Rho family GTPase 2 {[}Source:HGNC Symbol;Acc:HGNC:18315{]}          & ENSG00000108830 & 17 & 43025231 & 43032041 &  1\\\\\n",
       "\t chr17:43038208-43038434 & BRCA1 & BRCA1 DNA repair associated {[}Source:HGNC Symbol;Acc:HGNC:1100{]}   & ENSG00000012048 & 17 & 43044295 & 43170245 & -1\\\\\n",
       "\t chr17:43038558-43038759 & RND2  & Rho family GTPase 2 {[}Source:HGNC Symbol;Acc:HGNC:18315{]}          & ENSG00000108830 & 17 & 43025231 & 43032041 &  1\\\\\n",
       "\t chr17:43038558-43038759 & BRCA1 & BRCA1 DNA repair associated {[}Source:HGNC Symbol;Acc:HGNC:1100{]}   & ENSG00000012048 & 17 & 43044295 & 43170245 & -1\\\\\n",
       "\t chr17:43038933-43039159 & RND2  & Rho family GTPase 2 {[}Source:HGNC Symbol;Acc:HGNC:18315{]}          & ENSG00000108830 & 17 & 43025231 & 43032041 &  1\\\\\n",
       "\t chr17:43038933-43039159 & BRCA1 & BRCA1 DNA repair associated {[}Source:HGNC Symbol;Acc:HGNC:1100{]}   & ENSG00000012048 & 17 & 43044295 & 43170245 & -1\\\\\n",
       "\t chr17:43041533-43042034 & RND2  & Rho family GTPase 2 {[}Source:HGNC Symbol;Acc:HGNC:18315{]}          & ENSG00000108830 & 17 & 43025231 & 43032041 &  1\\\\\n",
       "\t chr17:43041533-43042034 & BRCA1 & BRCA1 DNA repair associated {[}Source:HGNC Symbol;Acc:HGNC:1100{]}   & ENSG00000012048 & 17 & 43044295 & 43170245 & -1\\\\\n",
       "\t chr17:43042908-43043509 & BRCA1 & BRCA1 DNA repair associated {[}Source:HGNC Symbol;Acc:HGNC:1100{]}   & ENSG00000012048 & 17 & 43044295 & 43170245 & -1\\\\\n",
       "\t chr17:43043708-43044559 & BRCA1 & BRCA1 DNA repair associated {[}Source:HGNC Symbol;Acc:HGNC:1100{]}   & ENSG00000012048 & 17 & 43044295 & 43170245 & -1\\\\\n",
       "\t chr17:43044783-43045709 & BRCA1 & BRCA1 DNA repair associated {[}Source:HGNC Symbol;Acc:HGNC:1100{]}   & ENSG00000012048 & 17 & 43044295 & 43170245 & -1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 35 √ó 8\n",
       "\n",
       "| site &lt;chr&gt; | hgnc_symbol &lt;chr&gt; | description &lt;chr&gt; | ensembl_gene_id &lt;chr&gt; | chromosome_name &lt;int&gt; | start_position &lt;int&gt; | end_position &lt;int&gt; | strand &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| chr17:43024283-43024634 | IFI35 | interferon induced protein 35 [Source:HGNC Symbol;Acc:HGNC:5399] | ENSG00000068079 | 17 | 43006740 | 43014456 |  1 |\n",
       "| chr17:43024283-43024634 | VAT1  | vesicle amine transport 1 [Source:HGNC Symbol;Acc:HGNC:16919]    | ENSG00000108828 | 17 | 43014607 | 43025123 | -1 |\n",
       "| chr17:43024283-43024634 | RND2  | Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]          | ENSG00000108830 | 17 | 43025231 | 43032041 |  1 |\n",
       "| chr17:43024658-43025384 | VAT1  | vesicle amine transport 1 [Source:HGNC Symbol;Acc:HGNC:16919]    | ENSG00000108828 | 17 | 43014607 | 43025123 | -1 |\n",
       "| chr17:43024658-43025384 | RND2  | Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]          | ENSG00000108830 | 17 | 43025231 | 43032041 |  1 |\n",
       "| chr17:43025608-43025834 | VAT1  | vesicle amine transport 1 [Source:HGNC Symbol;Acc:HGNC:16919]    | ENSG00000108828 | 17 | 43014607 | 43025123 | -1 |\n",
       "| chr17:43025608-43025834 | RND2  | Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]          | ENSG00000108830 | 17 | 43025231 | 43032041 |  1 |\n",
       "| chr17:43027833-43028184 | VAT1  | vesicle amine transport 1 [Source:HGNC Symbol;Acc:HGNC:16919]    | ENSG00000108828 | 17 | 43014607 | 43025123 | -1 |\n",
       "| chr17:43027833-43028184 | RND2  | Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]          | ENSG00000108830 | 17 | 43025231 | 43032041 |  1 |\n",
       "| chr17:43029258-43029509 | VAT1  | vesicle amine transport 1 [Source:HGNC Symbol;Acc:HGNC:16919]    | ENSG00000108828 | 17 | 43014607 | 43025123 | -1 |\n",
       "| chr17:43029258-43029509 | RND2  | Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]          | ENSG00000108830 | 17 | 43025231 | 43032041 |  1 |\n",
       "| chr17:43029733-43029959 | VAT1  | vesicle amine transport 1 [Source:HGNC Symbol;Acc:HGNC:16919]    | ENSG00000108828 | 17 | 43014607 | 43025123 | -1 |\n",
       "| chr17:43029733-43029959 | RND2  | Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]          | ENSG00000108830 | 17 | 43025231 | 43032041 |  1 |\n",
       "| chr17:43033283-43034184 | VAT1  | vesicle amine transport 1 [Source:HGNC Symbol;Acc:HGNC:16919]    | ENSG00000108828 | 17 | 43014607 | 43025123 | -1 |\n",
       "| chr17:43033283-43034184 | RND2  | Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]          | ENSG00000108830 | 17 | 43025231 | 43032041 |  1 |\n",
       "| chr17:43034208-43034834 | VAT1  | vesicle amine transport 1 [Source:HGNC Symbol;Acc:HGNC:16919]    | ENSG00000108828 | 17 | 43014607 | 43025123 | -1 |\n",
       "| chr17:43034208-43034834 | RND2  | Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]          | ENSG00000108830 | 17 | 43025231 | 43032041 |  1 |\n",
       "| chr17:43034208-43034834 | BRCA1 | BRCA1 DNA repair associated [Source:HGNC Symbol;Acc:HGNC:1100]   | ENSG00000012048 | 17 | 43044295 | 43170245 | -1 |\n",
       "| chr17:43035633-43036034 | RND2  | Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]          | ENSG00000108830 | 17 | 43025231 | 43032041 |  1 |\n",
       "| chr17:43035633-43036034 | BRCA1 | BRCA1 DNA repair associated [Source:HGNC Symbol;Acc:HGNC:1100]   | ENSG00000012048 | 17 | 43044295 | 43170245 | -1 |\n",
       "| chr17:43037183-43037409 | RND2  | Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]          | ENSG00000108830 | 17 | 43025231 | 43032041 |  1 |\n",
       "| chr17:43037183-43037409 | BRCA1 | BRCA1 DNA repair associated [Source:HGNC Symbol;Acc:HGNC:1100]   | ENSG00000012048 | 17 | 43044295 | 43170245 | -1 |\n",
       "| chr17:43037658-43038184 | RND2  | Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]          | ENSG00000108830 | 17 | 43025231 | 43032041 |  1 |\n",
       "| chr17:43037658-43038184 | BRCA1 | BRCA1 DNA repair associated [Source:HGNC Symbol;Acc:HGNC:1100]   | ENSG00000012048 | 17 | 43044295 | 43170245 | -1 |\n",
       "| chr17:43038208-43038434 | RND2  | Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]          | ENSG00000108830 | 17 | 43025231 | 43032041 |  1 |\n",
       "| chr17:43038208-43038434 | BRCA1 | BRCA1 DNA repair associated [Source:HGNC Symbol;Acc:HGNC:1100]   | ENSG00000012048 | 17 | 43044295 | 43170245 | -1 |\n",
       "| chr17:43038558-43038759 | RND2  | Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]          | ENSG00000108830 | 17 | 43025231 | 43032041 |  1 |\n",
       "| chr17:43038558-43038759 | BRCA1 | BRCA1 DNA repair associated [Source:HGNC Symbol;Acc:HGNC:1100]   | ENSG00000012048 | 17 | 43044295 | 43170245 | -1 |\n",
       "| chr17:43038933-43039159 | RND2  | Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]          | ENSG00000108830 | 17 | 43025231 | 43032041 |  1 |\n",
       "| chr17:43038933-43039159 | BRCA1 | BRCA1 DNA repair associated [Source:HGNC Symbol;Acc:HGNC:1100]   | ENSG00000012048 | 17 | 43044295 | 43170245 | -1 |\n",
       "| chr17:43041533-43042034 | RND2  | Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]          | ENSG00000108830 | 17 | 43025231 | 43032041 |  1 |\n",
       "| chr17:43041533-43042034 | BRCA1 | BRCA1 DNA repair associated [Source:HGNC Symbol;Acc:HGNC:1100]   | ENSG00000012048 | 17 | 43044295 | 43170245 | -1 |\n",
       "| chr17:43042908-43043509 | BRCA1 | BRCA1 DNA repair associated [Source:HGNC Symbol;Acc:HGNC:1100]   | ENSG00000012048 | 17 | 43044295 | 43170245 | -1 |\n",
       "| chr17:43043708-43044559 | BRCA1 | BRCA1 DNA repair associated [Source:HGNC Symbol;Acc:HGNC:1100]   | ENSG00000012048 | 17 | 43044295 | 43170245 | -1 |\n",
       "| chr17:43044783-43045709 | BRCA1 | BRCA1 DNA repair associated [Source:HGNC Symbol;Acc:HGNC:1100]   | ENSG00000012048 | 17 | 43044295 | 43170245 | -1 |\n",
       "\n"
      ],
      "text/plain": [
       "   site                    hgnc_symbol\n",
       "1  chr17:43024283-43024634 IFI35      \n",
       "2  chr17:43024283-43024634 VAT1       \n",
       "3  chr17:43024283-43024634 RND2       \n",
       "4  chr17:43024658-43025384 VAT1       \n",
       "5  chr17:43024658-43025384 RND2       \n",
       "6  chr17:43025608-43025834 VAT1       \n",
       "7  chr17:43025608-43025834 RND2       \n",
       "8  chr17:43027833-43028184 VAT1       \n",
       "9  chr17:43027833-43028184 RND2       \n",
       "10 chr17:43029258-43029509 VAT1       \n",
       "11 chr17:43029258-43029509 RND2       \n",
       "12 chr17:43029733-43029959 VAT1       \n",
       "13 chr17:43029733-43029959 RND2       \n",
       "14 chr17:43033283-43034184 VAT1       \n",
       "15 chr17:43033283-43034184 RND2       \n",
       "16 chr17:43034208-43034834 VAT1       \n",
       "17 chr17:43034208-43034834 RND2       \n",
       "18 chr17:43034208-43034834 BRCA1      \n",
       "19 chr17:43035633-43036034 RND2       \n",
       "20 chr17:43035633-43036034 BRCA1      \n",
       "21 chr17:43037183-43037409 RND2       \n",
       "22 chr17:43037183-43037409 BRCA1      \n",
       "23 chr17:43037658-43038184 RND2       \n",
       "24 chr17:43037658-43038184 BRCA1      \n",
       "25 chr17:43038208-43038434 RND2       \n",
       "26 chr17:43038208-43038434 BRCA1      \n",
       "27 chr17:43038558-43038759 RND2       \n",
       "28 chr17:43038558-43038759 BRCA1      \n",
       "29 chr17:43038933-43039159 RND2       \n",
       "30 chr17:43038933-43039159 BRCA1      \n",
       "31 chr17:43041533-43042034 RND2       \n",
       "32 chr17:43041533-43042034 BRCA1      \n",
       "33 chr17:43042908-43043509 BRCA1      \n",
       "34 chr17:43043708-43044559 BRCA1      \n",
       "35 chr17:43044783-43045709 BRCA1      \n",
       "   description                                                     \n",
       "1  interferon induced protein 35 [Source:HGNC Symbol;Acc:HGNC:5399]\n",
       "2  vesicle amine transport 1 [Source:HGNC Symbol;Acc:HGNC:16919]   \n",
       "3  Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]         \n",
       "4  vesicle amine transport 1 [Source:HGNC Symbol;Acc:HGNC:16919]   \n",
       "5  Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]         \n",
       "6  vesicle amine transport 1 [Source:HGNC Symbol;Acc:HGNC:16919]   \n",
       "7  Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]         \n",
       "8  vesicle amine transport 1 [Source:HGNC Symbol;Acc:HGNC:16919]   \n",
       "9  Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]         \n",
       "10 vesicle amine transport 1 [Source:HGNC Symbol;Acc:HGNC:16919]   \n",
       "11 Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]         \n",
       "12 vesicle amine transport 1 [Source:HGNC Symbol;Acc:HGNC:16919]   \n",
       "13 Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]         \n",
       "14 vesicle amine transport 1 [Source:HGNC Symbol;Acc:HGNC:16919]   \n",
       "15 Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]         \n",
       "16 vesicle amine transport 1 [Source:HGNC Symbol;Acc:HGNC:16919]   \n",
       "17 Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]         \n",
       "18 BRCA1 DNA repair associated [Source:HGNC Symbol;Acc:HGNC:1100]  \n",
       "19 Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]         \n",
       "20 BRCA1 DNA repair associated [Source:HGNC Symbol;Acc:HGNC:1100]  \n",
       "21 Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]         \n",
       "22 BRCA1 DNA repair associated [Source:HGNC Symbol;Acc:HGNC:1100]  \n",
       "23 Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]         \n",
       "24 BRCA1 DNA repair associated [Source:HGNC Symbol;Acc:HGNC:1100]  \n",
       "25 Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]         \n",
       "26 BRCA1 DNA repair associated [Source:HGNC Symbol;Acc:HGNC:1100]  \n",
       "27 Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]         \n",
       "28 BRCA1 DNA repair associated [Source:HGNC Symbol;Acc:HGNC:1100]  \n",
       "29 Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]         \n",
       "30 BRCA1 DNA repair associated [Source:HGNC Symbol;Acc:HGNC:1100]  \n",
       "31 Rho family GTPase 2 [Source:HGNC Symbol;Acc:HGNC:18315]         \n",
       "32 BRCA1 DNA repair associated [Source:HGNC Symbol;Acc:HGNC:1100]  \n",
       "33 BRCA1 DNA repair associated [Source:HGNC Symbol;Acc:HGNC:1100]  \n",
       "34 BRCA1 DNA repair associated [Source:HGNC Symbol;Acc:HGNC:1100]  \n",
       "35 BRCA1 DNA repair associated [Source:HGNC Symbol;Acc:HGNC:1100]  \n",
       "   ensembl_gene_id chromosome_name start_position end_position strand\n",
       "1  ENSG00000068079 17              43006740       43014456      1    \n",
       "2  ENSG00000108828 17              43014607       43025123     -1    \n",
       "3  ENSG00000108830 17              43025231       43032041      1    \n",
       "4  ENSG00000108828 17              43014607       43025123     -1    \n",
       "5  ENSG00000108830 17              43025231       43032041      1    \n",
       "6  ENSG00000108828 17              43014607       43025123     -1    \n",
       "7  ENSG00000108830 17              43025231       43032041      1    \n",
       "8  ENSG00000108828 17              43014607       43025123     -1    \n",
       "9  ENSG00000108830 17              43025231       43032041      1    \n",
       "10 ENSG00000108828 17              43014607       43025123     -1    \n",
       "11 ENSG00000108830 17              43025231       43032041      1    \n",
       "12 ENSG00000108828 17              43014607       43025123     -1    \n",
       "13 ENSG00000108830 17              43025231       43032041      1    \n",
       "14 ENSG00000108828 17              43014607       43025123     -1    \n",
       "15 ENSG00000108830 17              43025231       43032041      1    \n",
       "16 ENSG00000108828 17              43014607       43025123     -1    \n",
       "17 ENSG00000108830 17              43025231       43032041      1    \n",
       "18 ENSG00000012048 17              43044295       43170245     -1    \n",
       "19 ENSG00000108830 17              43025231       43032041      1    \n",
       "20 ENSG00000012048 17              43044295       43170245     -1    \n",
       "21 ENSG00000108830 17              43025231       43032041      1    \n",
       "22 ENSG00000012048 17              43044295       43170245     -1    \n",
       "23 ENSG00000108830 17              43025231       43032041      1    \n",
       "24 ENSG00000012048 17              43044295       43170245     -1    \n",
       "25 ENSG00000108830 17              43025231       43032041      1    \n",
       "26 ENSG00000012048 17              43044295       43170245     -1    \n",
       "27 ENSG00000108830 17              43025231       43032041      1    \n",
       "28 ENSG00000012048 17              43044295       43170245     -1    \n",
       "29 ENSG00000108830 17              43025231       43032041      1    \n",
       "30 ENSG00000012048 17              43044295       43170245     -1    \n",
       "31 ENSG00000108830 17              43025231       43032041      1    \n",
       "32 ENSG00000012048 17              43044295       43170245     -1    \n",
       "33 ENSG00000012048 17              43044295       43170245     -1    \n",
       "34 ENSG00000012048 17              43044295       43170245     -1    \n",
       "35 ENSG00000012048 17              43044295       43170245     -1    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#cell 34\n",
    "mapgenes <- NULL\n",
    "for (i in 1:nrow(info)) {\n",
    "  genes <-getBM(attributes = c(\"hgnc_symbol\", \"description\", \"ensembl_gene_id\",\"chromosome_name\",\n",
    "                              \"start_position\", \"end_position\", \"strand\"),\n",
    "               filters = c(\"chromosome_name\", \"start\", \"end\"),\n",
    "               values = list(gsub(\"chr\", \"\", info$chr[i]),info$start[i] - 10000, info$end[i] + 10000), mart = ensembl)\n",
    "  genes$site <- paste0(info$chr[i], \":\", info$start[i], \"-\", info$end[i])\n",
    "  mapgenes <- rbind(mapgenes, genes)\n",
    "}\n",
    "mapgenes[,c(8, 1:7)] # pour redoronner les colonnes et voir les resultats"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\"><b>Success:</b> Well done! Vous savez maintenant installer des paquets, indetifier les librairies o√π ils sont install√©s et utilisez deux paquets embl√©matiques en g√©nomique. \n",
    "</div>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
